<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tools for IMC data analysis • imcRtools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tools for IMC data analysis">
<meta property="og:description" content="imcRtools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">imcRtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/imcRtools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BodenmillerGroup/imcRtools/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="imcRtools_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="imcRtools_files/htmlwidgets-1.5.4/htmlwidgets.js"></script><link href="imcRtools_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="imcRtools_files/datatables-binding-0.20/datatables.js"></script><script src="imcRtools_files/jquery-3.6.0/jquery-3.6.0.min.js"></script><link href="imcRtools_files/dt-core-1.11.3/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="imcRtools_files/dt-core-1.11.3/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="imcRtools_files/dt-core-1.11.3/js/jquery.dataTables.min.js"></script><link href="imcRtools_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="imcRtools_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tools for IMC data analysis</h1>
                        <h4 class="author">Nils Eling</h4>
            <address class="author_afil">
      Department for Quantitative Biomedicine, University of Zurich<br><a class="author_email" href="mailto:#"></a><a href="mailto:nils.eling@dqbm.uzh.ch" class="email">nils.eling@dqbm.uzh.ch</a>
      </address>
                  
            <h4 class="date">28 November 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BodenmillerGroup/imcRtools/blob/master/vignettes/imcRtools.Rmd"><code>vignettes/imcRtools.Rmd</code></a></small>
      <div class="hidden name"><code>imcRtools.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This R package supports the handling and analysis of imaging mass cytometry and other highly multiplexed imaging data. The main functionality includes reading in single-cell data after image segmentation and measurement, data formatting to perform channel spillover correction and a number of spatial analysis approaches. First, cell-cell interactions are detected via spatial graph construction; these graphs can be visualized with cells representing nodes and interactions representing edges. Furthermore, per cell, its direct neighbours are summarized to allow spatial clustering. Per image/grouping level, interactions between types of cells are counted, averaged and compared against random permutations. In that way, types of cells that interact more (attraction) or less (avoidance) frequently than expected by chance are detected.</p>
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This vignette gives an introduction to handling and analyzing imaging mass cytometry (IMC) and other highly-multiplexed imaging data in R. The <code>imcRtools</code> package relies on expression and morphological features extracted from multi-channel images using corresponding segmentation masks. A description of data types and segmentation approaches can be found below (<a href="#dataTypes">data types</a>, <a href="#segmentation">segmentation</a>). However, due to shared data structures, the functionalities of the <code>imcRtools</code> package are applicable to most highly multiplexed imaging modalities.</p>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>The <code>imcRtools</code> package exports functions and example data to perform the following analyses:</p>
<ol style="list-style-type: decimal">
<li><a href="#read-in-data">Read in pre-processed data</a></li>
<li><a href="#spillover">Perform spillover correction for IMC data</a></li>
<li>
<a href="#build-graph">Build</a> and <a href="#viz-cells">visualize</a> spatial graphs</li>
<li>
<a href="#aggregate-neighbors">Aggregate across neighbouring cells</a> for spatial clustering</li>
<li>
<a href="#patch-detection">Detect spatial patches</a> of similar cell-types</li>
<li><a href="#test-neighborhood">Test the attraction or avoidance between celltypes</a></li>
</ol>
<p>To highlight the usability of these function, the <code>imcRtools</code> package also exports a number of <a href="#test-data">test data files</a>.</p>
</div>
<div id="highly-multiplexed-imaging" class="section level2">
<h2 class="hasAnchor">
<a href="#highly-multiplexed-imaging" class="anchor"></a>Highly multiplexed imaging</h2>
<p>Highly multiplexed imaging techniques <span class="citation">(Bodenmiller 2016)</span> such as imaging mass cytometry (IMC) <span class="citation">(Giesen et al. 2014)</span>, multiplexed ion beam imaging (MIBI) <span class="citation">(Angelo et al. 2014)</span> and cyclic immunofluorescence techniques <span class="citation">(Lin et al. 2018, @Gut2018)</span> acquire read-outs of the expression of tens of protein in a spatially resolved manner. After technology-dependent data pre-processing, the raw data files are comparable: multi-channel images for the dimensions <code>x</code>, <code>y</code>, and <code>c</code>, where <code>x</code> and <code>y</code> define the number of pixels (<code>x * y</code>) per image and <code>c</code> the number of proteins (also refered to as “markers”) measured per image. More info on the <a href="#dataTypes">data types</a> and <a href="#segmentation">further pre-processing</a> can be found below.</p>
<p>Increased multiplexity compared to epitope-based techniques is achieved using single-cell resolved spatial transcriptomics techniques including MERFISH <span class="citation">(Chen et al. 2015)</span> and seqFISH <span class="citation">(Lubeck et al. 2014)</span>. However, data produced by these techniques required different pre-processing steps. Nevertheless, analysis performed on the single-cell level is equally applicable.</p>
<div id="imaging-mass-cytometry" class="section level3">
<h3 class="hasAnchor">
<a href="#imaging-mass-cytometry" class="anchor"></a>Imaging mass cytometry</h3>
<p>IMC <span class="citation">(Giesen et al. 2014)</span> is a highly multiplexed imaging approach to measure spatial protein abundance. In IMC, tissue sections on slides are stained with a mix of around 40 metal-conjugated antibodies prior to laser ablation with <span class="math inline">\(1\mu{}m\)</span> resolution. The ablated material is transferred to a mass cytometer for time-of-flight detection of the metal ions <span class="citation">(Giesen et al. 2014)</span>. At an ablation frequency of 200Hz, a 1mm x 1mm region can be acquired within 1.5 hours.</p>
</div>
<div id="dataTypes" class="section level3">
<h3 class="hasAnchor">
<a href="#dataTypes" class="anchor"></a>Data types</h3>
<p>In the case of IMC, the raw data output are .mcd files containing all acquired regions per slide. During image pre-prcoessing these files are converted into individual multi-channel .tiff and <a href="https://docs.openmicroscopy.org/ome-model/5.6.3/ome-tiff/">OME-TIFF</a> files. These file formats are supported by most open-source and commercial image analysis software, such as <a href="https://imagej.net/software/fiji/">Fiji</a>, <a href="https://qupath.github.io/">QuPath</a> or <a href="https://napari.org/">napari</a>.</p>
<p>In R, these .tiff files can be read into individual <a href="https://bioconductor.org/packages/release/bioc/vignettes/EBImage/inst/doc/EBImage-introduction.html#3_Image_data_representation">Image</a> objects or combined into a <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/cytomapper/inst/doc/cytomapper.html#5_The_CytoImageList_object">CytoImageList</a> object supported by the <a href="https://www.bioconductor.org/packages/release/bioc/html/cytomapper.html">cytomapper</a> package.</p>
</div>
<div id="segmentation" class="section level3">
<h3 class="hasAnchor">
<a href="#segmentation" class="anchor"></a>Segmentation and feature extraction</h3>
<p>The pixel resolution of most highly multiplexed imaging technologies including IMC support the detection of cellular structures. Therefore, a common processing step using multi-channel images is object segmentation. In this vignette, objects are defined as cells; however, also larger scale structures could be segmented.</p>
<p>There are multiple different image segmentation approaches available. However, <code>imcRtools</code> only supports direct reader functions for two segmentation strategies developed for highly multiplexed imaging technologies:</p>
<ol style="list-style-type: decimal">
<li><p>The <a href="https://github.com/BodenmillerGroup/ImcSegmentationPipeline">ImcSegmentationPipeline</a> has been developed to give the user flexibility in how to perform channel selection and image segmentation. The underlying principle is to train a pixel classifier (using <a href="https://www.ilastik.org/">ilastik</a>) on a number of selected channels to perform probabilistic classification of each pixel into: background, cytoplasm and nucleus. Based on these classification probabilities a <a href="https://cellprofiler.org/">CellProfiler</a> pipeline performs cell segmentation and feature extraction.</p></li>
<li><p>A containerized version of this pipeline is implemented in <a href="https://github.com/BodenmillerGroup/steinbock">steinbock</a>. <code>steinbock</code> further supports segmentation via the use of <code>Mesmer</code> from the <code>DeepCell</code> library <span class="citation">(Greenwald et al. 2021)</span>.</p></li>
</ol>
<p>While the output of both approaches is structured differently, the exported features are comparable:</p>
<ol style="list-style-type: decimal">
<li>per cell: channel intensity, morphology and location</li>
<li>cell-cell interactions exported as graph</li>
</ol>
<p>By combining these with availabel channel information, the data can be read into a <em><a href="https://bioconductor.org/packages/3.14/SpatialExperiment">SpatialExperiment</a></em> or <em><a href="https://bioconductor.org/packages/3.14/SingleCellExperiment">SingleCellExperiment</a></em> object (see <a href="#read-in-data">below</a>).</p>
</div>
</div>
</div>
<div id="test-data" class="section level1">
<h1 class="hasAnchor">
<a href="#test-data" class="anchor"></a>Example data</h1>
<p>The <code>imcRtools</code> package contains a number of example data generated by the Hyperion imaging system for different purposes. The following section gives an overview of these files.</p>
<div id="for-spillover-correction" class="section level2">
<h2 class="hasAnchor">
<a href="#for-spillover-correction" class="anchor"></a>For spillover correction</h2>
<p>To highlight the use of the <code>imcRtools</code> package for spillover correction, we provide four .txt files containing pixel intensities of four spotted metals.</p>
<p>Please refer to the <a href="#spillover">spillover correction</a> section for full details.</p>
<p>These files are accessible via:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/spillover"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">path</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "Dy161.txt" "Dy162.txt" "Dy163.txt" "Dy164.txt"</code></pre>
</div>
<div id="raw-data-in-form-of--txt-files" class="section level2">
<h2 class="hasAnchor">
<a href="#raw-data-in-form-of--txt-files" class="anchor"></a>Raw data in form of .txt files</h2>
<p>IMC generates .mcd files storing the raw data or all acquired regions of interest (ROI). In addition, the raw pixel values are also stored in individual .txt files per ROI.</p>
<p>To highlight reading in raw data in form of .txt files, the <code>imcRtools</code> contains 3 sample acquisitions:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">txt_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/mockData/raw"</span>, 
                                    package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span><span class="op">)</span>
<span class="va">txt_files</span></code></pre></div>
<pre><code>## [1] "20210305_NE_mockData2_ROI_001_1.txt" "20210305_NE_mockData2_ROI_002_2.txt"
## [3] "20210305_NE_mockData2_ROI_003_3.txt"</code></pre>
</div>
<div id="imcsegmentationpipeline-output-data" class="section level2">
<h2 class="hasAnchor">
<a href="#imcsegmentationpipeline-output-data" class="anchor"></a>ImcSegmentationPipeline output data</h2>
<p>IMC data preprocessing and segmentation can be performed using the <a href="https://github.com/BodenmillerGroup/ImcSegmentationPipeline">ImcSegmentationPipeline</a>. It generates a number of <code>.csv</code> files containing object/cell-specific and image-specific metadata.</p>
<p>The <code>imcRtools</code> package exports the <code>read_cpout</code> function as convenient reader function for outputs generated by the <code>ImcSegmentationPipeline</code>. For demonstration purposes, <code>imcRtools</code> contains the output of running the pipeline on a small example dataset:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/mockData/cpout"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">path</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "cell.csv"                 "Experiment.csv"          
## [3] "Image.csv"                "Object_relationships.csv"
## [5] "panel.csv"                "var_cell.csv"            
## [7] "var_Image.csv"</code></pre>
</div>
<div id="steinbock-output-data" class="section level2">
<h2 class="hasAnchor">
<a href="#steinbock-output-data" class="anchor"></a>steinbock output data</h2>
<p>The <a href="https://github.com/BodenmillerGroup/steinbock">steinbock</a> pipeline can be used to process, segment and extract features from IMC data. For more information, please refer to its <a href="https://bodenmillergroup.github.io/steinbock/">documentation</a>.</p>
<p>To highlight the functionality of <code>imcRtools</code> to read in single-cell data generated by <code>steinbock</code>, we provide a small toy dataset available at:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/mockData/steinbock"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">path</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "cells.csv"                              
##  [2] "images.csv"                             
##  [3] "intensities/20210305_NE_mockData1_1.csv"
##  [4] "intensities/20210305_NE_mockData1_2.csv"
##  [5] "intensities/20210305_NE_mockData1_3.csv"
##  [6] "intensities/20210305_NE_mockData2_1.csv"
##  [7] "intensities/20210305_NE_mockData2_2.csv"
##  [8] "intensities/20210305_NE_mockData2_3.csv"
##  [9] "intensities/20210305_NE_mockData3_1.csv"
## [10] "intensities/20210305_NE_mockData3_2.csv"
## [11] "intensities/20210305_NE_mockData3_3.csv"
## [12] "intensities/20210305_NE_mockData4_1.csv"
## [13] "intensities/20210305_NE_mockData4_2.csv"
## [14] "intensities/20210305_NE_mockData4_3.csv"
## [15] "intensities/20210305_NE_mockData5_1.csv"
## [16] "intensities/20210305_NE_mockData5_2.csv"
## [17] "intensities/20210305_NE_mockData5_3.csv"
## [18] "neighbors/20210305_NE_mockData1_1.csv"  
## [19] "neighbors/20210305_NE_mockData1_2.csv"  
## [20] "neighbors/20210305_NE_mockData1_3.csv"  
## [21] "neighbors/20210305_NE_mockData2_1.csv"  
## [22] "neighbors/20210305_NE_mockData2_2.csv"  
## [23] "neighbors/20210305_NE_mockData2_3.csv"  
## [24] "neighbors/20210305_NE_mockData3_1.csv"  
## [25] "neighbors/20210305_NE_mockData3_2.csv"  
## [26] "neighbors/20210305_NE_mockData3_3.csv"  
## [27] "neighbors/20210305_NE_mockData4_1.csv"  
## [28] "neighbors/20210305_NE_mockData4_2.csv"  
## [29] "neighbors/20210305_NE_mockData4_3.csv"  
## [30] "neighbors/20210305_NE_mockData5_1.csv"  
## [31] "neighbors/20210305_NE_mockData5_2.csv"  
## [32] "neighbors/20210305_NE_mockData5_3.csv"  
## [33] "panel.csv"                              
## [34] "regionprops/20210305_NE_mockData1_1.csv"
## [35] "regionprops/20210305_NE_mockData1_2.csv"
## [36] "regionprops/20210305_NE_mockData1_3.csv"
## [37] "regionprops/20210305_NE_mockData2_1.csv"
## [38] "regionprops/20210305_NE_mockData2_2.csv"
## [39] "regionprops/20210305_NE_mockData2_3.csv"
## [40] "regionprops/20210305_NE_mockData3_1.csv"
## [41] "regionprops/20210305_NE_mockData3_2.csv"
## [42] "regionprops/20210305_NE_mockData3_3.csv"
## [43] "regionprops/20210305_NE_mockData4_1.csv"
## [44] "regionprops/20210305_NE_mockData4_2.csv"
## [45] "regionprops/20210305_NE_mockData4_3.csv"
## [46] "regionprops/20210305_NE_mockData5_1.csv"
## [47] "regionprops/20210305_NE_mockData5_2.csv"
## [48] "regionprops/20210305_NE_mockData5_3.csv"
## [49] "steinbock.sh"</code></pre>
<p>The example data related to the <code>ImcSegmentationPipeline</code> and <code>steinbock</code> can also be accessed <a href="https://github.com/BodenmillerGroup/TestData/tree/main/datasets/210308_ImcTestData">online</a>.</p>
</div>
</div>
<div id="read-in-data" class="section level1">
<h1 class="hasAnchor">
<a href="#read-in-data" class="anchor"></a>Read in IMC data</h1>
<p>The <code>imcRtools</code> package supports reading in data generated by the <a href="https://github.com/BodenmillerGroup/ImcSegmentationPipeline">ImcSegmentationPipeline</a> or <a href="https://github.com/BodenmillerGroup/steinbock">steinbock</a> pipeline.</p>
<p>To read in the outpout data into a <em><a href="https://bioconductor.org/packages/3.14/SpatialExperiment">SpatialExperiment</a></em> or <em><a href="https://bioconductor.org/packages/3.14/SingleCellExperiment">SingleCellExperiment</a></em>, the <code>imcRtools</code> package exports the <code>read_cpout</code> function.</p>
<p>By default, the single-cell data is read into a <em><a href="https://bioconductor.org/packages/3.14/SpatialExperiment">SpatialExperiment</a></em> object. Here, the extracted channel- and cell-specific intensities are stored in the <code>counts(spe)</code> slot. All morphological features are stored in <code>colData(spe)</code> and the spatial locations of the cells are stored in <code>spatialCoords(spe)</code>. The interaction graph is stored in <code>colPair(spe, "neighbourhood")</code>.</p>
<p>Alternatively, the data can be read into a <em><a href="https://bioconductor.org/packages/3.14/SingleCellExperiment">SingleCellExperiment</a></em> object. The only difference is the lack of <code>spatialCoords(sce)</code>. Here, the spatial coordinates are stored in <code>colData(spe)$Pos_X</code> and <code>colData(spe)$Pos_Y</code>.</p>
<div id="read-in-cellprofiler-output" class="section level2">
<h2 class="hasAnchor">
<a href="#read-in-cellprofiler-output" class="anchor"></a>Read in CellProfiler output</h2>
<p>The <a href="https://github.com/BodenmillerGroup/ImcSegmentationPipeline">ImcSegmentationPipeline</a> produces a number of output files. By default, all single-cell features are measured and exported. To browse and select the features of interest, the <code>imcRtools</code> package provides the <code>show_cpout_features</code> function:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/mockData/cpout"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span>

<span class="fu"><a href="../reference/show_cpout_features.html">show_cpout_features</a></span><span class="op">(</span><span class="va">path</span><span class="op">)</span></code></pre></div>
<div id="htmlwidget-219fcc12ca7214a4453d" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-219fcc12ca7214a4453d">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200","201","202","203","204","205","206","207","208","209","210","211","212","213","214","215","216","217","218","219","220","221","222","223","224","225","226","227","228","229","230","231","232","233","234","235","236","237","238","239","240","241","242","243","244","245","246","247","248","249","250","251","252","253","254","255","256","257","258","259","260","261","262","263","264","265","266","267","268","269","270","271","272","273","274","275","276","277","278","279","280","281","282","283","284"],["Location_Center_X","Location_Center_Y","Location_Center_Z","Number_Object_Number","Neighbors_NumberOfNeighbors_8","Neighbors_PercentTouching_8","Neighbors_FirstClosestObjectNumber_8","Neighbors_FirstClosestDistance_8","Neighbors_SecondClosestObjectNumber_8","Neighbors_SecondClosestDistance_8","Neighbors_AngleBetweenNeighbors_8","Intensity_IntegratedIntensity_FullStack_c1","Intensity_MeanIntensity_FullStack_c1","Intensity_StdIntensity_FullStack_c1","Intensity_MinIntensity_FullStack_c1","Intensity_MaxIntensity_FullStack_c1","Intensity_IntegratedIntensityEdge_FullStack_c1","Intensity_MeanIntensityEdge_FullStack_c1","Intensity_StdIntensityEdge_FullStack_c1","Intensity_MinIntensityEdge_FullStack_c1","Intensity_MaxIntensityEdge_FullStack_c1","Intensity_MassDisplacement_FullStack_c1","Intensity_LowerQuartileIntensity_FullStack_c1","Intensity_MedianIntensity_FullStack_c1","Intensity_MADIntensity_FullStack_c1","Intensity_UpperQuartileIntensity_FullStack_c1","Location_CenterMassIntensity_X_FullStack_c1","Location_CenterMassIntensity_Y_FullStack_c1","Location_CenterMassIntensity_Z_FullStack_c1","Location_MaxIntensity_X_FullStack_c1","Location_MaxIntensity_Y_FullStack_c1","Location_MaxIntensity_Z_FullStack_c1","Intensity_IntegratedIntensity_FullStack_c2","Intensity_MeanIntensity_FullStack_c2","Intensity_StdIntensity_FullStack_c2","Intensity_MinIntensity_FullStack_c2","Intensity_MaxIntensity_FullStack_c2","Intensity_IntegratedIntensityEdge_FullStack_c2","Intensity_MeanIntensityEdge_FullStack_c2","Intensity_StdIntensityEdge_FullStack_c2","Intensity_MinIntensityEdge_FullStack_c2","Intensity_MaxIntensityEdge_FullStack_c2","Intensity_MassDisplacement_FullStack_c2","Intensity_LowerQuartileIntensity_FullStack_c2","Intensity_MedianIntensity_FullStack_c2","Intensity_MADIntensity_FullStack_c2","Intensity_UpperQuartileIntensity_FullStack_c2","Location_CenterMassIntensity_X_FullStack_c2","Location_CenterMassIntensity_Y_FullStack_c2","Location_CenterMassIntensity_Z_FullStack_c2","Location_MaxIntensity_X_FullStack_c2","Location_MaxIntensity_Y_FullStack_c2","Location_MaxIntensity_Z_FullStack_c2","Intensity_IntegratedIntensity_FullStack_c3","Intensity_MeanIntensity_FullStack_c3","Intensity_StdIntensity_FullStack_c3","Intensity_MinIntensity_FullStack_c3","Intensity_MaxIntensity_FullStack_c3","Intensity_IntegratedIntensityEdge_FullStack_c3","Intensity_MeanIntensityEdge_FullStack_c3","Intensity_StdIntensityEdge_FullStack_c3","Intensity_MinIntensityEdge_FullStack_c3","Intensity_MaxIntensityEdge_FullStack_c3","Intensity_MassDisplacement_FullStack_c3","Intensity_LowerQuartileIntensity_FullStack_c3","Intensity_MedianIntensity_FullStack_c3","Intensity_MADIntensity_FullStack_c3","Intensity_UpperQuartileIntensity_FullStack_c3","Location_CenterMassIntensity_X_FullStack_c3","Location_CenterMassIntensity_Y_FullStack_c3","Location_CenterMassIntensity_Z_FullStack_c3","Location_MaxIntensity_X_FullStack_c3","Location_MaxIntensity_Y_FullStack_c3","Location_MaxIntensity_Z_FullStack_c3","Intensity_IntegratedIntensity_FullStack_c4","Intensity_MeanIntensity_FullStack_c4","Intensity_StdIntensity_FullStack_c4","Intensity_MinIntensity_FullStack_c4","Intensity_MaxIntensity_FullStack_c4","Intensity_IntegratedIntensityEdge_FullStack_c4","Intensity_MeanIntensityEdge_FullStack_c4","Intensity_StdIntensityEdge_FullStack_c4","Intensity_MinIntensityEdge_FullStack_c4","Intensity_MaxIntensityEdge_FullStack_c4","Intensity_MassDisplacement_FullStack_c4","Intensity_LowerQuartileIntensity_FullStack_c4","Intensity_MedianIntensity_FullStack_c4","Intensity_MADIntensity_FullStack_c4","Intensity_UpperQuartileIntensity_FullStack_c4","Location_CenterMassIntensity_X_FullStack_c4","Location_CenterMassIntensity_Y_FullStack_c4","Location_CenterMassIntensity_Z_FullStack_c4","Location_MaxIntensity_X_FullStack_c4","Location_MaxIntensity_Y_FullStack_c4","Location_MaxIntensity_Z_FullStack_c4","Intensity_IntegratedIntensity_FullStack_c5","Intensity_MeanIntensity_FullStack_c5","Intensity_StdIntensity_FullStack_c5","Intensity_MinIntensity_FullStack_c5","Intensity_MaxIntensity_FullStack_c5","Intensity_IntegratedIntensityEdge_FullStack_c5","Intensity_MeanIntensityEdge_FullStack_c5","Intensity_StdIntensityEdge_FullStack_c5","Intensity_MinIntensityEdge_FullStack_c5","Intensity_MaxIntensityEdge_FullStack_c5","Intensity_MassDisplacement_FullStack_c5","Intensity_LowerQuartileIntensity_FullStack_c5","Intensity_MedianIntensity_FullStack_c5","Intensity_MADIntensity_FullStack_c5","Intensity_UpperQuartileIntensity_FullStack_c5","Location_CenterMassIntensity_X_FullStack_c5","Location_CenterMassIntensity_Y_FullStack_c5","Location_CenterMassIntensity_Z_FullStack_c5","Location_MaxIntensity_X_FullStack_c5","Location_MaxIntensity_Y_FullStack_c5","Location_MaxIntensity_Z_FullStack_c5","Intensity_IntegratedIntensity_FullStackFiltered_c1","Intensity_MeanIntensity_FullStackFiltered_c1","Intensity_StdIntensity_FullStackFiltered_c1","Intensity_MinIntensity_FullStackFiltered_c1","Intensity_MaxIntensity_FullStackFiltered_c1","Intensity_IntegratedIntensityEdge_FullStackFiltered_c1","Intensity_MeanIntensityEdge_FullStackFiltered_c1","Intensity_StdIntensityEdge_FullStackFiltered_c1","Intensity_MinIntensityEdge_FullStackFiltered_c1","Intensity_MaxIntensityEdge_FullStackFiltered_c1","Intensity_MassDisplacement_FullStackFiltered_c1","Intensity_LowerQuartileIntensity_FullStackFiltered_c1","Intensity_MedianIntensity_FullStackFiltered_c1","Intensity_MADIntensity_FullStackFiltered_c1","Intensity_UpperQuartileIntensity_FullStackFiltered_c1","Location_CenterMassIntensity_X_FullStackFiltered_c1","Location_CenterMassIntensity_Y_FullStackFiltered_c1","Location_CenterMassIntensity_Z_FullStackFiltered_c1","Location_MaxIntensity_X_FullStackFiltered_c1","Location_MaxIntensity_Y_FullStackFiltered_c1","Location_MaxIntensity_Z_FullStackFiltered_c1","Intensity_IntegratedIntensity_FullStackFiltered_c2","Intensity_MeanIntensity_FullStackFiltered_c2","Intensity_StdIntensity_FullStackFiltered_c2","Intensity_MinIntensity_FullStackFiltered_c2","Intensity_MaxIntensity_FullStackFiltered_c2","Intensity_IntegratedIntensityEdge_FullStackFiltered_c2","Intensity_MeanIntensityEdge_FullStackFiltered_c2","Intensity_StdIntensityEdge_FullStackFiltered_c2","Intensity_MinIntensityEdge_FullStackFiltered_c2","Intensity_MaxIntensityEdge_FullStackFiltered_c2","Intensity_MassDisplacement_FullStackFiltered_c2","Intensity_LowerQuartileIntensity_FullStackFiltered_c2","Intensity_MedianIntensity_FullStackFiltered_c2","Intensity_MADIntensity_FullStackFiltered_c2","Intensity_UpperQuartileIntensity_FullStackFiltered_c2","Location_CenterMassIntensity_X_FullStackFiltered_c2","Location_CenterMassIntensity_Y_FullStackFiltered_c2","Location_CenterMassIntensity_Z_FullStackFiltered_c2","Location_MaxIntensity_X_FullStackFiltered_c2","Location_MaxIntensity_Y_FullStackFiltered_c2","Location_MaxIntensity_Z_FullStackFiltered_c2","Intensity_IntegratedIntensity_FullStackFiltered_c3","Intensity_MeanIntensity_FullStackFiltered_c3","Intensity_StdIntensity_FullStackFiltered_c3","Intensity_MinIntensity_FullStackFiltered_c3","Intensity_MaxIntensity_FullStackFiltered_c3","Intensity_IntegratedIntensityEdge_FullStackFiltered_c3","Intensity_MeanIntensityEdge_FullStackFiltered_c3","Intensity_StdIntensityEdge_FullStackFiltered_c3","Intensity_MinIntensityEdge_FullStackFiltered_c3","Intensity_MaxIntensityEdge_FullStackFiltered_c3","Intensity_MassDisplacement_FullStackFiltered_c3","Intensity_LowerQuartileIntensity_FullStackFiltered_c3","Intensity_MedianIntensity_FullStackFiltered_c3","Intensity_MADIntensity_FullStackFiltered_c3","Intensity_UpperQuartileIntensity_FullStackFiltered_c3","Location_CenterMassIntensity_X_FullStackFiltered_c3","Location_CenterMassIntensity_Y_FullStackFiltered_c3","Location_CenterMassIntensity_Z_FullStackFiltered_c3","Location_MaxIntensity_X_FullStackFiltered_c3","Location_MaxIntensity_Y_FullStackFiltered_c3","Location_MaxIntensity_Z_FullStackFiltered_c3","Intensity_IntegratedIntensity_FullStackFiltered_c4","Intensity_MeanIntensity_FullStackFiltered_c4","Intensity_StdIntensity_FullStackFiltered_c4","Intensity_MinIntensity_FullStackFiltered_c4","Intensity_MaxIntensity_FullStackFiltered_c4","Intensity_IntegratedIntensityEdge_FullStackFiltered_c4","Intensity_MeanIntensityEdge_FullStackFiltered_c4","Intensity_StdIntensityEdge_FullStackFiltered_c4","Intensity_MinIntensityEdge_FullStackFiltered_c4","Intensity_MaxIntensityEdge_FullStackFiltered_c4","Intensity_MassDisplacement_FullStackFiltered_c4","Intensity_LowerQuartileIntensity_FullStackFiltered_c4","Intensity_MedianIntensity_FullStackFiltered_c4","Intensity_MADIntensity_FullStackFiltered_c4","Intensity_UpperQuartileIntensity_FullStackFiltered_c4","Location_CenterMassIntensity_X_FullStackFiltered_c4","Location_CenterMassIntensity_Y_FullStackFiltered_c4","Location_CenterMassIntensity_Z_FullStackFiltered_c4","Location_MaxIntensity_X_FullStackFiltered_c4","Location_MaxIntensity_Y_FullStackFiltered_c4","Location_MaxIntensity_Z_FullStackFiltered_c4","Intensity_IntegratedIntensity_FullStackFiltered_c5","Intensity_MeanIntensity_FullStackFiltered_c5","Intensity_StdIntensity_FullStackFiltered_c5","Intensity_MinIntensity_FullStackFiltered_c5","Intensity_MaxIntensity_FullStackFiltered_c5","Intensity_IntegratedIntensityEdge_FullStackFiltered_c5","Intensity_MeanIntensityEdge_FullStackFiltered_c5","Intensity_StdIntensityEdge_FullStackFiltered_c5","Intensity_MinIntensityEdge_FullStackFiltered_c5","Intensity_MaxIntensityEdge_FullStackFiltered_c5","Intensity_MassDisplacement_FullStackFiltered_c5","Intensity_LowerQuartileIntensity_FullStackFiltered_c5","Intensity_MedianIntensity_FullStackFiltered_c5","Intensity_MADIntensity_FullStackFiltered_c5","Intensity_UpperQuartileIntensity_FullStackFiltered_c5","Location_CenterMassIntensity_X_FullStackFiltered_c5","Location_CenterMassIntensity_Y_FullStackFiltered_c5","Location_CenterMassIntensity_Z_FullStackFiltered_c5","Location_MaxIntensity_X_FullStackFiltered_c5","Location_MaxIntensity_Y_FullStackFiltered_c5","Location_MaxIntensity_Z_FullStackFiltered_c5","Intensity_IntegratedIntensity_ProbSegmentation_c1","Intensity_MeanIntensity_ProbSegmentation_c1","Intensity_StdIntensity_ProbSegmentation_c1","Intensity_MinIntensity_ProbSegmentation_c1","Intensity_MaxIntensity_ProbSegmentation_c1","Intensity_IntegratedIntensityEdge_ProbSegmentation_c1","Intensity_MeanIntensityEdge_ProbSegmentation_c1","Intensity_StdIntensityEdge_ProbSegmentation_c1","Intensity_MinIntensityEdge_ProbSegmentation_c1","Intensity_MaxIntensityEdge_ProbSegmentation_c1","Intensity_MassDisplacement_ProbSegmentation_c1","Intensity_LowerQuartileIntensity_ProbSegmentation_c1","Intensity_MedianIntensity_ProbSegmentation_c1","Intensity_MADIntensity_ProbSegmentation_c1","Intensity_UpperQuartileIntensity_ProbSegmentation_c1","Location_CenterMassIntensity_X_ProbSegmentation_c1","Location_CenterMassIntensity_Y_ProbSegmentation_c1","Location_CenterMassIntensity_Z_ProbSegmentation_c1","Location_MaxIntensity_X_ProbSegmentation_c1","Location_MaxIntensity_Y_ProbSegmentation_c1","Location_MaxIntensity_Z_ProbSegmentation_c1","Intensity_IntegratedIntensity_ProbSegmentation_c2","Intensity_MeanIntensity_ProbSegmentation_c2","Intensity_StdIntensity_ProbSegmentation_c2","Intensity_MinIntensity_ProbSegmentation_c2","Intensity_MaxIntensity_ProbSegmentation_c2","Intensity_IntegratedIntensityEdge_ProbSegmentation_c2","Intensity_MeanIntensityEdge_ProbSegmentation_c2","Intensity_StdIntensityEdge_ProbSegmentation_c2","Intensity_MinIntensityEdge_ProbSegmentation_c2","Intensity_MaxIntensityEdge_ProbSegmentation_c2","Intensity_MassDisplacement_ProbSegmentation_c2","Intensity_LowerQuartileIntensity_ProbSegmentation_c2","Intensity_MedianIntensity_ProbSegmentation_c2","Intensity_MADIntensity_ProbSegmentation_c2","Intensity_UpperQuartileIntensity_ProbSegmentation_c2","Location_CenterMassIntensity_X_ProbSegmentation_c2","Location_CenterMassIntensity_Y_ProbSegmentation_c2","Location_CenterMassIntensity_Z_ProbSegmentation_c2","Location_MaxIntensity_X_ProbSegmentation_c2","Location_MaxIntensity_Y_ProbSegmentation_c2","Location_MaxIntensity_Z_ProbSegmentation_c2","Intensity_IntegratedIntensity_ProbSegmentation_c3","Intensity_MeanIntensity_ProbSegmentation_c3","Intensity_StdIntensity_ProbSegmentation_c3","Intensity_MinIntensity_ProbSegmentation_c3","Intensity_MaxIntensity_ProbSegmentation_c3","Intensity_IntegratedIntensityEdge_ProbSegmentation_c3","Intensity_MeanIntensityEdge_ProbSegmentation_c3","Intensity_StdIntensityEdge_ProbSegmentation_c3","Intensity_MinIntensityEdge_ProbSegmentation_c3","Intensity_MaxIntensityEdge_ProbSegmentation_c3","Intensity_MassDisplacement_ProbSegmentation_c3","Intensity_LowerQuartileIntensity_ProbSegmentation_c3","Intensity_MedianIntensity_ProbSegmentation_c3","Intensity_MADIntensity_ProbSegmentation_c3","Intensity_UpperQuartileIntensity_ProbSegmentation_c3","Location_CenterMassIntensity_X_ProbSegmentation_c3","Location_CenterMassIntensity_Y_ProbSegmentation_c3","Location_CenterMassIntensity_Z_ProbSegmentation_c3","Location_MaxIntensity_X_ProbSegmentation_c3","Location_MaxIntensity_Y_ProbSegmentation_c3","Location_MaxIntensity_Z_ProbSegmentation_c3"],["Location","Location","Location","ID","Neighbors","Neighbors","Neighbors","Neighbors","Neighbors","Neighbors","Neighbors","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location"],[null,null,null,null,null,null,null,null,null,null,null,"FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","FullStackFiltered","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation","ProbSegmentation"],[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],["Center_X","Center_Y","Center_Z","Number_Object_Number","NumberOfNeighbors","PercentTouching","FirstClosestObjectNumber","FirstClosestDistance","SecondClosestObjectNumber","SecondClosestDistance","AngleBetweenNeighbors","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z"],[null,null,null,null,null,null,null,null,null,null,null,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3],[null,null,null,null,8,8,8,8,8,8,8,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],[null,null,null,null,null,null,null,null,null,null,null,"Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background"],["float","float","float","integer","integer","float","integer","float","integer","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>column_name<\/th>\n      <th>category<\/th>\n      <th>image_name<\/th>\n      <th>object_name<\/th>\n      <th>feature_name<\/th>\n      <th>channel<\/th>\n      <th>parameters<\/th>\n      <th>channel_id<\/th>\n      <th>data_type<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[6,7]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>By default, <code>read_cpout</code> will read in the mean intensity per channel and cell from “hot pixel” filtered image stacks specified via <code>intensities = "Intensity_MeanIntensity_FullStackFiltered"</code>. Please refer to <code><a href="../reference/read_cpout.html">?read_cpout</a></code> for the full documentation.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cur_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/mockData/cpout"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span>

<span class="co"># Read as SpatialExperiment</span>
<span class="op">(</span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_cpout.html">read_cpout</a></span><span class="op">(</span><span class="va">cur_path</span>, graph_file <span class="op">=</span> <span class="st">"Object_relationships.csv"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## class: SpatialExperiment 
## dim: 5 219 
## metadata(0):
## assays(1): counts
## rownames(5): Ag107 Pr141 Sm147 Eu153 Yb172
## rowData names(5): Tube.Number Metal.Tag Target ilastik full
## colnames: NULL
## colData names(6): sample_id ObjectNumber ... Metadata_acid
##   Metadata_description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## spatialData names(0) :
## spatialCoords names(2) : Pos_X Pos_Y
## imgData names(1): sample_id</code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Read as SingleCellExperiment</span>
<span class="op">(</span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_cpout.html">read_cpout</a></span><span class="op">(</span><span class="va">cur_path</span>, graph_file <span class="op">=</span> <span class="st">"Object_relationships.csv"</span>, 
                   return_as <span class="op">=</span> <span class="st">"sce"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 5 219 
## metadata(0):
## assays(1): counts
## rownames(5): Ag107 Pr141 Sm147 Eu153 Yb172
## rowData names(5): Tube.Number Metal.Tag Target ilastik full
## colnames: NULL
## colData names(8): sample_id ObjectNumber ... Metadata_acid
##   Metadata_description
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):</code></pre>
</div>
<div id="read-in-steinbock-output" class="section level2">
<h2 class="hasAnchor">
<a href="#read-in-steinbock-output" class="anchor"></a>Read in steinbock output</h2>
<p>Single-cell data and all associated metadata (e.g. spatial location, morphology and interaction graphs) as produced by the <a href="https://github.com/BodenmillerGroup/steinbock">steinbock</a> pipeline can be read in using the <code>read_steinbock</code> function:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cur_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/mockData/steinbock"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span>

<span class="co"># Read as SpatialExperiment</span>
<span class="op">(</span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_steinbock.html">read_steinbock</a></span><span class="op">(</span><span class="va">cur_path</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## class: SpatialExperiment 
## dim: 5 218 
## metadata(0):
## assays(1): counts
## rownames(5): Ag107 Cytokeratin 5 Laminin YBX1 H3K27Ac
## rowData names(6): channel name ... deepcell Tube.Number
## colnames: NULL
## colData names(8): sample_id ObjectNumber ... width_px height_px
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):
## spatialData names(0) :
## spatialCoords names(2) : Pos_X Pos_Y
## imgData names(1): sample_id</code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Read as SingleCellExperiment</span>
<span class="op">(</span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_steinbock.html">read_steinbock</a></span><span class="op">(</span><span class="va">cur_path</span>, return_as <span class="op">=</span> <span class="st">"sce"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 5 218 
## metadata(0):
## assays(1): counts
## rownames(5): Ag107 Cytokeratin 5 Laminin YBX1 H3K27Ac
## rowData names(6): channel name ... deepcell Tube.Number
## colnames: NULL
## colData names(10): sample_id ObjectNumber ... width_px height_px
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):</code></pre>
<p>For more information, please refer to <code><a href="../reference/read_steinbock.html">?read_steinbock</a></code>.</p>
</div>
<div id="read-raw--txt-files-into-image-objects" class="section level2">
<h2 class="hasAnchor">
<a href="#read-raw--txt-files-into-image-objects" class="anchor"></a>Read raw .txt files into Image objects</h2>
<p>For reading in and visualization of multi-channel images and segmentation masks, please refer to the <em><a href="https://bioconductor.org/packages/3.14/cytomapper">cytomapper</a></em> package. The <code>imcRtools</code> package however supports reading in raw .txt files generated by the Hyperion imaging system into a <code>CytoImageList</code> object; a data container exported by <code>cytomapper</code>.</p>
<p>The user needs to provide a path from which all .txt files will be read in:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/mockData/raw"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span>

<span class="va">cur_CytoImageList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readImagefromTXT.html">readImagefromTXT</a></span><span class="op">(</span><span class="va">path</span><span class="op">)</span>
<span class="va">cur_CytoImageList</span></code></pre></div>
<pre><code>## CytoImageList containing 3 image(s)
## names(3): 20210305_NE_mockData2_ROI_001_1 20210305_NE_mockData2_ROI_002_2 20210305_NE_mockData2_ROI_003_3 
## Each image contains 5 channel(s)
## channelNames(5): Ag107Di Pr141Di Sm147Di Eu153Di Yb172Di</code></pre>
<p>By specifying the <code>pattern</code> argument, individual or a subset of files can be read in. For more information, please refer to <code><a href="../reference/readImagefromTXT.html">?readImagefromTXT</a></code>.</p>
</div>
</div>
<div id="spillover" class="section level1">
<h1 class="hasAnchor">
<a href="#spillover" class="anchor"></a>Spillover correction</h1>
<p>When acquiring IMC images, pixel intensities can be influenced by spillover from neighboring channels. To correct for this, Chevrier <em>et al.</em> have developed a staining protocol to acquire individually spotted metal isotopes <span class="citation">(Chevrier et al. 2017)</span>. Based on these measurements, spillover into neighboring channels can be quantified to correct pixel intensities.</p>
<p>The <code>imcRtools</code> package provides helper functions that facilitate the correction of spillover for IMC data. For a full tutorial, please refer to the <a href="https://bodenmillergroup.github.io/IMCDataAnalysis/spillover-correction.html">IMC data analysis book</a>.</p>
<div id="read-in-the-single-spot-acquisitions" class="section level2">
<h2 class="hasAnchor">
<a href="#read-in-the-single-spot-acquisitions" class="anchor"></a>Read in the single-spot acquisitions</h2>
<p>In the first step, the pixel intensities of individually spotted metals need to be read into a <code>SingleCellExperiment</code> container for downstream use with the <em><a href="https://bioconductor.org/packages/3.14/CATALYST">CATALYST</a></em> package. For this, the <code>readSCEfromTXT</code> function can be used:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/spillover"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readSCEfromTXT.html">readSCEfromTXT</a></span><span class="op">(</span><span class="va">path</span><span class="op">)</span> </code></pre></div>
<pre><code>## Spotted channels:  Dy161, Dy162, Dy163, Dy164
## Acquired channels:  Dy161, Dy162, Dy163, Dy164
## Channels spotted but not acquired:  
## Channels acquired but not spotted:</code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 4 400 
## metadata(0):
## assays(1): counts
## rownames(4): 161Dy(Dy161Di) 162Dy(Dy162Di) 163Dy(Dy163Di)
##   164Dy(Dy164Di)
## rowData names(2): channel_name marker_name
## colnames(400): Dy161.1 Dy161.2 ... Dy164.99 Dy164.100
## colData names(9): Start_push End_push ... sample_metal sample_mass
## reducedDimNames(0):
## mainExpName: NULL
## altExpNames(0):</code></pre>
<p>Here, the example metal spot files are read in. The spot information are stored in the <code>colData(sce)</code> slot and channel information are stored in <code>rowData(sce)</code>. Each column represents a single pixel.</p>
</div>
<div id="quality-control-on-single-spot-acquisitions" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-control-on-single-spot-acquisitions" class="anchor"></a>Quality control on single-spot acquisitions</h2>
<p>In the next step, it is crucial to identify potentially mislabeled spots or spots with low pixel intensities. The <code>imcRtools</code> package exports the <code>plotSpotHeatmap</code> function, which visualizes the aggregated (default <code>median</code>) pixel intensities per spot and per metal:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSpotHeatmap.html">plotSpotHeatmap</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></code></pre></div>
<p><img src="imcRtools_files/figure-html/plotSpotHeatmap-1.png" width="480"></p>
<p>Here, high median pixel intensities can be observed in each spot and their corresponding channels (visualized on the <code>log10</code> scale by default). To quickly identify spot/channel combinations with low signal, the <code>threshold</code> parameter can be set:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotSpotHeatmap.html">plotSpotHeatmap</a></span><span class="op">(</span><span class="va">sce</span>, log <span class="op">=</span> <span class="cn">FALSE</span>, threshold <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></code></pre></div>
<p><img src="imcRtools_files/figure-html/plotSpotHeatmap-2-1.png" width="480"></p>
</div>
<div id="consecutive-pixel-binning" class="section level2">
<h2 class="hasAnchor">
<a href="#consecutive-pixel-binning" class="anchor"></a>Consecutive pixel binning</h2>
<p>If pixel intensities are low, spillover estimation might not be robust. Therefore, the <code>binAcrossPixels</code> function can be used to sum consecutive pixels and enhance the acquired signal. This step is optional for spillover estimation.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/binAcrossPixels.html">binAcrossPixels</a></span><span class="op">(</span><span class="va">sce</span>, bin_size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotSpotHeatmap.html">plotSpotHeatmap</a></span><span class="op">(</span><span class="va">sce2</span>, log <span class="op">=</span> <span class="cn">FALSE</span>, threshold <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></code></pre></div>
<p><img src="imcRtools_files/figure-html/pixel-binning-1.png" width="480"></p>
</div>
<div id="pixel-filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#pixel-filtering" class="anchor"></a>Pixel filtering</h2>
<p>Prior to spillover estimation, the <em><a href="https://bioconductor.org/packages/3.14/CATALYST">CATALYST</a></em> package provides the <code>assignPrelim</code>, <code>estCutoffs</code> and <code>applyCutoffs</code> functions to estimate the spotted mass for each pixel based on their channel intensities. For more information on the spillover estimation and correction, please refer to the <a href="https://bioconductor.org/packages/release/bioc/vignettes/CATALYST/inst/doc/preprocessing.html#compensation">CATALYST vignette</a>.</p>
<p>This estimation can be used to identify pixels that cannot be easily assigned to their spotted mass, potentially indicating pixels with weak signal. To remove these pixels, the <code>filterPixels</code> function can be used. This function further removes pixels assigned to masses, which only contain very few pixels.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/CATALYST">CATALYST</a></span><span class="op">)</span>

<span class="va">bc_key</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">sample_mass</span><span class="op">)</span><span class="op">)</span>
<span class="fu">assay</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"exprs"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Hyperbolic.html">asinh</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">/</span><span class="fl">5</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CATALYST/man/assignPrelim.html">assignPrelim</a></span><span class="op">(</span><span class="va">sce</span>, bc_key <span class="op">=</span> <span class="va">bc_key</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CATALYST/man/estCutoffs.html">estCutoffs</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CATALYST/man/applyCutoffs.html">applyCutoffs</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>

<span class="co"># Filter out mislabeled pixels</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterPixels.html">filterPixels</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">bc_id</span>, <span class="va">sce</span><span class="op">$</span><span class="va">sample_mass</span><span class="op">)</span></code></pre></div>
<pre><code>##      
##       161 162 163 164
##   0     0   0   0   2
##   161 100   0   0   0
##   162   0 100   0   0
##   163   0   0 100   0
##   164   0   0   0  98</code></pre>
</div>
<div id="estimating-the-spillover-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#estimating-the-spillover-matrix" class="anchor"></a>Estimating the spillover matrix</h2>
<p>Finally, the pre-processed <code>SiingleCellExperiment</code> object can be used to generate the spillover matrix using the <code><a href="https://rdrr.io/pkg/CATALYST/man/computeSpillmat.html">CATALYST::computeSpillmat</a></code> function:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CATALYST/man/computeSpillmat.html">computeSpillmat</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="fu">metadata</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">spillover_matrix</span></code></pre></div>
<pre><code>##             Dy161Di     Dy162Di     Dy163Di     Dy164Di
## Dy161Di 1.000000000 0.031443129 0.009734712 0.006518048
## Dy162Di 0.015715159 1.000000000 0.048116187 0.008250039
## Dy163Di 0.003809504 0.012159704 1.000000000 0.020214651
## Dy164Di 0.005058069 0.008457546 0.028912343 1.000000000</code></pre>
<p>This spillover matrix can be directly applied to compensate the summarized pixel intensities per cell and per channel as described <a href="https://bioconductor.org/packages/release/bioc/vignettes/CATALYST/inst/doc/preprocessing.html#compcytof-compensation-of-mass-cytometry-data">here</a>.</p>
</div>
</div>
<div id="spatial-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#spatial-analysis" class="anchor"></a>Spatial analysis</h1>
<p>The following section will highlight functions for spatial analyses of the data.</p>
<div id="build-graph" class="section level2">
<h2 class="hasAnchor">
<a href="#build-graph" class="anchor"></a>Constructing graphs</h2>
<p>When following the <code>ImcSegmentationPipeline</code> or <code>steinbock</code> and reading in the data using the corresponding functions, the generated graphs are automatically stored in the <code>colPair(spe, "neighborhood")</code> slot. Alternatively, the <code>buildSpatialGraph</code> function in the <code>imcRtools</code> package constructs interaction graphs using either (i) cell-centroid expansion, (ii) k-nearest neighbor search or (iii) delaunay triangulation.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BodenmillerGroup/cytomapper">cytomapper</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/CATALYST/man/data.html">data</a></span><span class="op">(</span><span class="st">"pancreasSCE"</span><span class="op">)</span>

<span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildSpatialGraph.html">buildSpatialGraph</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,
                                 type <span class="op">=</span> <span class="st">"expansion"</span>,
                                 threshold <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>
<span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildSpatialGraph.html">buildSpatialGraph</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,
                                 type <span class="op">=</span> <span class="st">"knn"</span>,
                                 k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>
<span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildSpatialGraph.html">buildSpatialGraph</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,
                                 type <span class="op">=</span> <span class="st">"delaunay"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colPairs.html">colPairNames</a></span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "expansion_interaction_graph" "knn_interaction_graph"      
## [3] "delaunay_interaction_graph"</code></pre>
<p>When setting <code>type = "knn"</code>, by default a directional graph will be build. Setting <code>directed = FALSE</code> will create bi-directional edges for each pair of cells that are connected by at least one edge in the directed setting.</p>
</div>
<div id="viz-cells" class="section level2">
<h2 class="hasAnchor">
<a href="#viz-cells" class="anchor"></a>Graph/cell visualization</h2>
<p>The cells’ locations and constructed graphs can be visualized using the <code>plotSpatial</code> function. Here, cells are referred to as “nodes” and cell-cell interactions are referred to as “edges”. All visual attributes of the nodes and edges can be set. Either by specifying a variable in <code>colData(spe)</code>, a marker name or a single entry using the <code>*_fix</code> parameters.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggraph.data-imaginist.com">ggraph</a></span><span class="op">)</span>

<span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,
            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,
            node_color_by <span class="op">=</span> <span class="st">"CellType"</span>,
            node_shape_by <span class="op">=</span> <span class="st">"ImageNb"</span>,
            node_size_by <span class="op">=</span> <span class="st">"Area"</span>,
            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,
            colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,
            directed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="imcRtools_files/figure-html/plotSpatial-1.png" width="700"></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Colored by expression and with arrows</span>
<span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,
            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,
            node_color_by <span class="op">=</span> <span class="st">"PIN"</span>,
            assay_type <span class="op">=</span> <span class="st">"exprs"</span>,
            node_size_fix <span class="op">=</span> <span class="fl">3</span>,
            edge_width_fix <span class="op">=</span> <span class="fl">0.2</span>,
            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,
            colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,
            directed <span class="op">=</span> <span class="cn">TRUE</span>,
            arrow <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/arrow.html">arrow</a></span><span class="op">(</span>length <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="st">"inch"</span><span class="op">)</span><span class="op">)</span>,
            end_cap <span class="op">=</span> <span class="fu">ggraph</span><span class="fu">::</span><span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geometry.html">circle</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="imcRtools_files/figure-html/plotSpatial-2.png" width="700"></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Subsetting the SingleCellExperiment</span>
<span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">[</span>,<span class="va">pancreasSCE</span><span class="op">$</span><span class="va">Pattern</span><span class="op">]</span>,
            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,
            node_color_by <span class="op">=</span> <span class="st">"CellType"</span>,
            node_size_fix <span class="op">=</span> <span class="fl">1</span>,
            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,
            colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,
            directed <span class="op">=</span> <span class="cn">TRUE</span>,
            scales <span class="op">=</span> <span class="st">"fixed"</span><span class="op">)</span> </code></pre></div>
<p><img src="imcRtools_files/figure-html/plotSpatial-3.png" width="700"></p>
<p>The returned object can be further modified using the <code>ggplot2</code> logic. This includes changing the node color, shape and size using <code>scale_color_*</code>, <code>scale_shape_*</code> and <code>scale_size_*</code>. Edge attributes can be altered using the <code>scale_edge_*</code> function exported by <code>ggraph</code>,</p>
</div>
<div id="aggregate-neighbors" class="section level2">
<h2 class="hasAnchor">
<a href="#aggregate-neighbors" class="anchor"></a>Neighborhood aggregation</h2>
<p>The <code>aggregateNeighbors</code> function can be used to aggregate features of all neighboring cells for each individual cell. This function operates in two settings.</p>
<ol style="list-style-type: decimal">
<li>
<code>metadata</code>: when aggregating by cell-specific metadata, the function computes the relative frequencies of all entries to <code>colData(sce)[[count_by]]</code> within the direct neighborhood of each cell.<br>
</li>
<li>
<code>expression</code>: the expression counts of neighboring cells are aggregated using the specified <code>statistic</code> (defaults to <code>mean</code>).</li>
</ol>
<p>Each cell’s neighborhood is defined as endpoints of edges stored in <code>colPair(sce, colPairName)</code>.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregateNeighbors.html">aggregateNeighbors</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,
                                  colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,
                                  aggregate_by <span class="op">=</span> <span class="st">"metadata"</span>,
                                  count_by <span class="op">=</span> <span class="st">"CellType"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">$</span><span class="va">aggregatedNeighbors</span><span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 3 columns
##   celltype_A celltype_B celltype_C
##    &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt;
## 1          0        0.0        1.0
## 2          0        0.2        0.8
## 3          0        0.0        1.0
## 4          0        0.0        1.0
## 5          0        0.0        1.0
## 6          0        0.0        1.0</code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregateNeighbors.html">aggregateNeighbors</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,
                                  colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,
                                  aggregate_by <span class="op">=</span> <span class="st">"expression"</span>,
                                  assay_type <span class="op">=</span>  <span class="st">"exprs"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">$</span><span class="va">mean_aggregatedExpression</span><span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 5 columns
##          H3      CD99       PIN      CD8a       CDH
##   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
## 1   2.32500  0.860329  0.092871  0.725000   2.51264
## 2   2.88022  1.629762  0.319527  0.207873   2.46486
## 3   3.10829  0.735389  0.190616  0.255515   1.89484
## 4   2.55842  0.773342  0.124545  0.188629   2.51084
## 5   2.44287  1.126240  0.252129  0.200261   2.61336
## 6   2.65059  0.903869  0.181792  0.196691   2.16434</code></pre>
<p>The returned entries can now be used for clustering to group cells based on their environment (either by aggregated categorical features or expression).</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cur_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html">kmeans</a></span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">$</span><span class="va">aggregatedNeighbors</span>, centers <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="va">pancreasSCE</span><span class="op">$</span><span class="va">clustered_neighbors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">cur_cluster</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,
            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,
            node_color_by <span class="op">=</span> <span class="st">"CellType"</span>,
            node_size_fix <span class="op">=</span> <span class="fl">4</span>,
            edge_width_fix <span class="op">=</span> <span class="fl">2</span>,
            edge_color_by <span class="op">=</span> <span class="st">"clustered_neighbors"</span>,
            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,
            colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,
            directed <span class="op">=</span> <span class="cn">FALSE</span>,
            nodes_first <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set2"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_colour.html">scale_edge_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></code></pre></div>
<p><img src="imcRtools_files/figure-html/aggregateNeigbors-clustering-1.png" width="700"></p>
<p>To exclude cells that are close to the image border, the <code>imcRtools</code> package exports the <code>findBorderCells</code> function.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findBorderCells.html">findBorderCells</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, 
                               img_id <span class="op">=</span> <span class="st">"ImageNb"</span>, 
                               border_dist <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">[</span>,<span class="op">!</span><span class="va">pancreasSCE</span><span class="op">$</span><span class="va">border_cells</span><span class="op">]</span>,
            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,
            node_color_by <span class="op">=</span> <span class="st">"CellType"</span>,
            node_size_fix <span class="op">=</span> <span class="fl">4</span>,
            edge_width_fix <span class="op">=</span> <span class="fl">2</span>,
            edge_color_by <span class="op">=</span> <span class="st">"clustered_neighbors"</span>,
            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,
            colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,
            directed <span class="op">=</span> <span class="cn">FALSE</span>,
            nodes_first <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set2"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_colour.html">scale_edge_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></code></pre></div>
<p><img src="imcRtools_files/figure-html/findBorderCells-1.png" width="700"></p>
<p>Excluding border cells can be useful when incorrectly connected cells are observed at image borders.</p>
</div>
<div id="patch-detection" class="section level2">
<h2 class="hasAnchor">
<a href="#patch-detection" class="anchor"></a>Patch detection</h2>
<p>An alternative and supervised way of detecting regions with similar types of cells is available via the <code>patchDetection</code> function. Here, the user defines which cells should be used for patch detection via the <code>patch_cells</code> parameter. A patch is defined as a set of cells as defined by <code>patch_cells</code> which are weakly connected in the graph in <code>colPair(object, colPairname)</code>.</p>
<p>Below, the <code>patchDetection</code> function is demonstrated using the previously computed <code>expansion</code> graph and defining cells of <code>celltype_B</code> as the cells of interest. Here, the function additionally draws a concave hull around the detected patch, expands the hull by 20<span class="math inline">\(\mu{}m\)</span> and defines all cells within this expanded hulls as patch cells.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/patchDetection.html">patchDetection</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, 
                              patch_cells <span class="op">=</span> <span class="va">pancreasSCE</span><span class="op">$</span><span class="va">CellType</span> <span class="op">==</span> <span class="st">"celltype_B"</span>,
                              colPairName <span class="op">=</span> <span class="st">"expansion_interaction_graph"</span>,
                              expand_by <span class="op">=</span> <span class="fl">20</span>, 
                              img_id <span class="op">=</span> <span class="st">"ImageNb"</span><span class="op">)</span>
 
<span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, img_id <span class="op">=</span> <span class="st">"ImageNb"</span>, node_color_by <span class="op">=</span> <span class="st">"patch_id"</span><span class="op">)</span></code></pre></div>
<p><img src="imcRtools_files/figure-html/patchDetection-1.png" width="700"></p>
<p>Patches that only consist of 1 or 2 cells cannot be expanded.</p>
</div>
<div id="test-neighborhood" class="section level2">
<h2 class="hasAnchor">
<a href="#test-neighborhood" class="anchor"></a>Neighborhood permutation testing</h2>
<p>The following section describes how to observe and test the average number of interactions between cell labels (e.g. cell-types) within grouping levels (e.g. images). For full descriptions of the testing approaches, please refer to <a href="https://www.nature.com/articles/nmeth.4391">Shapiro et al., Nature Methods</a> <span class="citation">(Schapiro et al. 2017)</span> and <a href="https://www.sciencedirect.com/science/article/pii/S2405471217305434">Schulz et al., Cell Systems</a> <span class="citation">(Schulz et al. 2018)</span></p>
<p>The <code>imcRtools</code> package exports the <code>countInteractions</code> and <code>testInteractions</code> functions, which summarize all cell-cell interactions per grouping level (e.g. image). As a result, a table is returned where each row represents one of all possible cell-type/cell-type interactions among all grouping levels. Missing entries or <code>NA</code>s indicate missing cell-type labels for this grouping level. The next section gives details on how interactions are summarized.</p>
<div id="summarizing-interactions" class="section level3">
<h3 class="hasAnchor">
<a href="#summarizing-interactions" class="anchor"></a>Summarizing interactions</h3>
<p>The <code>countInteractions</code> function counts the number of edges (interactions) between each set of unique cell labels per grouping level. Simplified, it counts for each cell of type A the number of neighbors of type B. This count is averaged within each unique grouping level (e.g. image) in three different ways:</p>
<ol style="list-style-type: decimal">
<li><p><code>method = "classic"</code>: The count is divided by the total number of cells of type A. The final count can be interpreted as “How many neighbors of type B does a cell of type A have on average?”</p></li>
<li><p><code>method = "histocat"</code>: The count is divided by the number of cells of type A that have at least one neighbor of type B. The final count can be interpreted as “How many many neighbors of type B has a cell of type A on average, given it has at least one neighbor of type B?”</p></li>
<li><p><code>method = "patch"</code>: For each cell, the count is binarized to 0 (less than <code>patch_size</code> neighbors of type B) or 1 (more or equal to <code>patch_size</code> neighbors of type B). The binarized counts are averaged across all cells of type A. The final count can be interpreted as “What fraction of cells of type A have at least a given number of neighbors of type B?”</p></li>
</ol>
<p>The <code>countInteractions</code> returns a <code>DataFrame</code> containing the summarized counts (<code>ct</code>) for all combinations of <code>from_label</code>, <code>to_label</code> and <code>group_by</code>.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/countInteractions.html">countInteractions</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,
                         group_by <span class="op">=</span> <span class="st">"ImageNb"</span>,
                         label <span class="op">=</span> <span class="st">"CellType"</span>,
                         method <span class="op">=</span> <span class="st">"classic"</span>,
                         colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span><span class="op">)</span>
<span class="va">out</span></code></pre></div>
<pre><code>## DataFrame with 27 rows and 4 columns
##      group_by from_label   to_label        ct
##     &lt;integer&gt;   &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt;
## 1           1 celltype_A celltype_A  2.823529
## 2           1 celltype_A celltype_B  0.823529
## 3           1 celltype_A celltype_C  1.352941
## 4           1 celltype_B celltype_A  2.000000
## 5           1 celltype_B celltype_B  0.625000
## ...       ...        ...        ...       ...
## 23          3 celltype_B celltype_B   4.00000
## 24          3 celltype_B celltype_C   1.00000
## 25          3 celltype_C celltype_A        NA
## 26          3 celltype_C celltype_B   1.13115
## 27          3 celltype_C celltype_C   3.86885</code></pre>
</div>
<div id="testing-for-significance" class="section level3">
<h3 class="hasAnchor">
<a href="#testing-for-significance" class="anchor"></a>Testing for significance</h3>
<p>In the next instance, one can test if the obtained count is larger or smaller compared to what is expected from a random distribution of cell labels. For this, the <code>testInteractions</code> function permutes the cell labels <code>iter</code> times and counts interactions as described above. This approach generates a distribution of the interaction count under a random distribution of cell labels. The observed interaction count is compared against this Null distribution to derive empirical p-values:</p>
<p><code>p_gt</code>: fraction of perturbations equal or greater than the observed count</p>
<p><code>p_lt</code>: fraction of perturbations equal or less than the observed count</p>
<p>Based on these empirical p-values, the <code>interaction</code> score (attraction or avoidance), overall <code>p</code> value and significance by comparison to <code>p_treshold</code> (<code>sig</code> and <code>sigval</code>) are derived. All results are returned in form of a <code>DataFrame</code>.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/testInteractions.html">testInteractions</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,
                        group_by <span class="op">=</span> <span class="st">"ImageNb"</span>,
                        label <span class="op">=</span> <span class="st">"CellType"</span>,
                        method <span class="op">=</span> <span class="st">"classic"</span>,
                        colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span><span class="op">)</span>
<span class="va">out</span></code></pre></div>
<pre><code>## DataFrame with 27 rows and 10 columns
##      group_by from_label   to_label        ct        p_gt        p_lt
##     &lt;integer&gt;   &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;
## 1           1 celltype_A celltype_A  2.823529 0.000999001 1.000000000
## 2           1 celltype_A celltype_B  0.823529 0.001998002 1.000000000
## 3           1 celltype_A celltype_C  1.352941 1.000000000 0.000999001
## 4           1 celltype_B celltype_A  2.000000 0.000999001 1.000000000
## 5           1 celltype_B celltype_B  0.625000 0.139860140 0.918081918
## ...       ...        ...        ...       ...         ...         ...
## 23          3 celltype_B celltype_B   4.00000 0.000999001 1.000000000
## 24          3 celltype_B celltype_C   1.00000 1.000000000 0.000999001
## 25          3 celltype_C celltype_A        NA          NA          NA
## 26          3 celltype_C celltype_B   1.13115 1.000000000 0.000999001
## 27          3 celltype_C celltype_C   3.86885 0.000999001 1.000000000
##     interaction           p       sig    sigval
##       &lt;logical&gt;   &lt;numeric&gt; &lt;logical&gt; &lt;numeric&gt;
## 1          TRUE 0.000999001      TRUE         1
## 2          TRUE 0.001998002      TRUE         1
## 3         FALSE 0.000999001      TRUE        -1
## 4          TRUE 0.000999001      TRUE         1
## 5          TRUE 0.139860140     FALSE         0
## ...         ...         ...       ...       ...
## 23         TRUE 0.000999001      TRUE         1
## 24        FALSE 0.000999001      TRUE        -1
## 25           NA          NA        NA        NA
## 26        FALSE 0.000999001      TRUE        -1
## 27         TRUE 0.000999001      TRUE         1</code></pre>
</div>
</div>
</div>
<div id="contributions" class="section level1">
<h1 class="hasAnchor">
<a href="#contributions" class="anchor"></a>Contributions</h1>
<p>Large chunks of the code of the <code>imcRtools</code> package is based on the work of <a href="https://github.com/votti">Vito Zanotelli</a>. Vito has co-developed the spillover correction approach and translated the interaction testing approach developed by <a href="https://github.com/DenisSch">Denis Shapiro</a> and <a href="https://github.com/JanaFischer">Jana Fischer</a> into R (formerly known as the <a href="https://github.com/BodenmillerGroup/neighbouRhood">neighbouRhood</a> R package). Jana has furthermore added the “patch” method for interaction counting and testing. <a href="https://github.com/toobiwankenobi">Tobias Hoch</a> has written the first version of reading in the <code>ImcSegmentationPipeline</code> output and the <code>patchDetection</code> function. <a href="https://github.com/SchulzDan">Daniel Schulz</a> has build the <code>aggregateNeighbors</code> function and contributed to developing the spatial clustering approach.</p>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Catalina 10.15.7
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] ggraph_2.0.5                ggplot2_3.3.5              
##  [3] cytomapper_1.4.1            EBImage_4.34.0             
##  [5] CATALYST_1.16.2             imcRtools_1.1.2            
##  [7] SpatialExperiment_1.2.1     SingleCellExperiment_1.14.1
##  [9] SummarizedExperiment_1.22.0 Biobase_2.52.0             
## [11] GenomicRanges_1.44.0        GenomeInfoDb_1.28.4        
## [13] IRanges_2.26.0              S4Vectors_0.30.2           
## [15] BiocGenerics_0.38.0         MatrixGenerics_1.4.3       
## [17] matrixStats_0.61.0          BiocStyle_2.20.2           
## 
## loaded via a namespace (and not attached):
##   [1] scattermore_0.7             flowWorkspace_4.4.0        
##   [3] R.methodsS3_1.8.1           ragg_1.2.0                 
##   [5] tidyr_1.1.4                 bit64_4.0.5                
##   [7] knitr_1.36                  irlba_2.3.3                
##   [9] multcomp_1.4-17             DelayedArray_0.18.0        
##  [11] R.utils_2.11.0              data.table_1.14.2          
##  [13] RCurl_1.98-1.5              doParallel_1.0.16          
##  [15] generics_0.1.1              flowCore_2.4.0             
##  [17] ScaledMatrix_1.0.0          TH.data_1.1-0              
##  [19] cowplot_1.1.1               terra_1.4-22               
##  [21] proxy_0.4-26                ggpointdensity_0.1.0       
##  [23] bit_4.0.4                   tzdb_0.2.0                 
##  [25] xml2_1.3.2                  httpuv_1.6.3               
##  [27] assertthat_0.2.1            viridis_0.6.2              
##  [29] xfun_0.28                   hms_1.1.1                  
##  [31] jquerylib_0.1.4             evaluate_0.14              
##  [33] promises_1.2.0.1            fansi_0.5.0                
##  [35] Rgraphviz_2.36.0            igraph_1.2.9               
##  [37] DBI_1.1.1                   htmlwidgets_1.5.4          
##  [39] purrr_0.3.4                 ellipsis_0.3.2             
##  [41] crosstalk_1.2.0             dplyr_1.0.7                
##  [43] ggcyto_1.20.0               ggnewscale_0.4.5           
##  [45] ggpubr_0.4.0                backports_1.4.0            
##  [47] V8_3.6.0                    cytolib_2.4.0              
##  [49] bookdown_0.24               svgPanZoom_0.3.4           
##  [51] RcppParallel_5.1.4          sparseMatrixStats_1.4.2    
##  [53] vctrs_0.3.8                 Cairo_1.5-12.2             
##  [55] abind_1.4-5                 withr_2.4.2                
##  [57] cachem_1.0.6                ggforce_0.3.3              
##  [59] aws.signature_0.6.0         vroom_1.5.6                
##  [61] svglite_2.0.0               cluster_2.1.2              
##  [63] crayon_1.4.2                drc_3.0-1                  
##  [65] labeling_0.4.2              edgeR_3.34.1               
##  [67] pkgconfig_2.0.3             units_0.7-2                
##  [69] tweenr_1.0.2                vipor_0.4.5                
##  [71] rlang_0.4.12                lifecycle_1.0.1            
##  [73] sandwich_3.0-1              rsvd_1.0.5                 
##  [75] rprojroot_2.0.2             polyclip_1.10-0            
##  [77] graph_1.70.0                tiff_0.1-10                
##  [79] Matrix_1.3-3                raster_3.5-2               
##  [81] carData_3.0-4               zoo_1.8-9                  
##  [83] Rhdf5lib_1.14.2             base64enc_0.1-3            
##  [85] beeswarm_0.4.0              RTriangle_1.6-0.10         
##  [87] ggridges_0.5.3              GlobalOptions_0.1.2        
##  [89] pheatmap_1.0.12             png_0.1-7                  
##  [91] viridisLite_0.4.0           rjson_0.2.20               
##  [93] bitops_1.0-7                shinydashboard_0.7.2       
##  [95] R.oo_1.24.0                 ConsensusClusterPlus_1.56.0
##  [97] KernSmooth_2.23-20          rhdf5filters_1.4.0         
##  [99] DelayedMatrixStats_1.14.3   shape_1.4.6                
## [101] classInt_0.4-3              stringr_1.4.0              
## [103] readr_2.1.0                 jpeg_0.1-9                 
## [105] rstatix_0.7.0               ggsignif_0.6.3             
## [107] aws.s3_0.3.21               beachmat_2.8.1             
## [109] scales_1.1.1                memoise_2.0.1              
## [111] magrittr_2.0.1              plyr_1.8.6                 
## [113] hexbin_1.28.2               zlibbioc_1.38.0            
## [115] compiler_4.1.0              dqrng_0.3.0                
## [117] concaveman_1.1.0            RColorBrewer_1.1-2         
## [119] plotrix_3.8-2               clue_0.3-60                
## [121] XVector_0.32.0              ncdfFlow_2.38.0            
## [123] FlowSOM_2.0.0               MASS_7.3-54                
## [125] tidyselect_1.1.1            stringi_1.7.5              
## [127] RProtoBufLib_2.4.0          textshaping_0.3.6          
## [129] highr_0.9                   yaml_2.2.1                 
## [131] BiocSingular_1.8.1          locfit_1.5-9.4             
## [133] latticeExtra_0.6-29         ggrepel_0.9.1              
## [135] grid_4.1.0                  sass_0.4.0                 
## [137] tools_4.1.0                 CytoML_2.4.0               
## [139] circlize_0.4.13             foreach_1.5.1              
## [141] gridExtra_2.3               farver_2.1.0               
## [143] Rtsne_0.15                  DropletUtils_1.12.3        
## [145] digest_0.6.28               BiocManager_1.30.16        
## [147] shiny_1.7.1                 Rcpp_1.0.7                 
## [149] car_3.0-12                  broom_0.7.10               
## [151] scuttle_1.2.1               later_1.3.0                
## [153] httr_1.4.2                  sf_1.0-4                   
## [155] ComplexHeatmap_2.8.0        colorspace_2.0-2           
## [157] XML_3.99-0.8                fs_1.5.0                   
## [159] splines_4.1.0               RBGL_1.68.0                
## [161] scater_1.20.1               pkgdown_1.6.1              
## [163] graphlayouts_0.7.2          sp_1.4-6                   
## [165] systemfonts_1.0.3           xtable_1.8-4               
## [167] jsonlite_1.7.2              tidygraph_1.2.0            
## [169] R6_2.5.1                    pillar_1.6.4               
## [171] htmltools_0.5.2             mime_0.12                  
## [173] nnls_1.4                    glue_1.5.0                 
## [175] fastmap_1.1.0               DT_0.20                    
## [177] BiocParallel_1.26.2         BiocNeighbors_1.10.0       
## [179] fftwtools_0.9-11            class_7.3-19               
## [181] codetools_0.2-18            mvtnorm_1.1-3              
## [183] utf8_1.2.2                  lattice_0.20-44            
## [185] bslib_0.3.1                 tibble_3.1.6               
## [187] curl_4.3.2                  ggbeeswarm_0.6.0           
## [189] colorRamps_2.3              gtools_3.9.2               
## [191] magick_2.7.3                survival_3.2-11            
## [193] limma_3.48.3                rmarkdown_2.11             
## [195] desc_1.4.0                  munsell_0.5.0              
## [197] e1071_1.7-9                 GetoptLong_1.0.5           
## [199] rhdf5_2.36.0                GenomeInfoDbData_1.2.6     
## [201] iterators_1.0.13            HDF5Array_1.20.0           
## [203] reshape2_1.4.4              gtable_0.3.0</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Angelo2014">
<p>Angelo, Michael, Sean C. Bendall, Rachel Finck, Matthew B. Hale, Chuck Hitzman, Alexander D. Borowsky, Richard M. Levenson, et al. 2014. “Multiplexed Ion Beam Imaging of Human Breast Tumors.” <em>Nature Medicine</em> 20 (4): 436–42.</p>
</div>
<div id="ref-Bodenmiller2016">
<p>Bodenmiller, Bernd. 2016. “Multiplexed Epitope-Based Tissue Imaging for Discovery and Healthcare Applications.” <em>Cell Systems</em> 2: 225–38.</p>
</div>
<div id="ref-Chen2015">
<p>Chen, Kok Hao, Alistair N. Boettiger, Jeffrey R. Moffitt, Siyuan Wang, and Xiaowei Zhuang. 2015. “Spatially Resolved, Highly Multiplexed Rna Profiling in Single Cells.” <em>Science</em> 348: aaa6090.</p>
</div>
<div id="ref-Chevrier2017">
<p>Chevrier, Stéphane, Helena L. Crowell, Vito R. T. Zanotelli, Stefanie Engler, Mark D. Robinson, and Bernd Bodenmiller. 2017. “Compensation of Signal Spillover in Suspension and Imaging Mass Cytometry.” <em>Cell Systems</em> 6: 612–20.</p>
</div>
<div id="ref-Giesen2014">
<p>Giesen, Charlotte, Hao A. O. Wang, Denis Schapiro, Nevena Zivanovic, Andrea Jacobs, Bodo Hattendorf, Peter J. Schüffler, et al. 2014. “Highly Multiplexed Imaging of Tumor Tissues with Subcellular Resolution by Mass Cytometry.” <em>Nature Methods</em> 11 (4): 417–22.</p>
</div>
<div id="ref-Greenwald2021">
<p>Greenwald, Noah F, Geneva Miller, Erick Moen, Alex Kong, Adam Kagel, Christine Camacho, Brianna J Mcintosh, et al. 2021. “Whole-Cell Segmentation of Tissue Images with Human-Level Performance Using Large-Scale Data Annotation and Deep Learning.” <em>bioRxiv</em>, 1–29.</p>
</div>
<div id="ref-Gut2018">
<p>Gut, Gabriele, Markus D Herrmann, and Lucas Pelkmans. 2018. “Multiplexed Protein Maps Link Subcellular Organization to Cellular States.” <em>Science</em> 361: 1–13.</p>
</div>
<div id="ref-Lin2018">
<p>Lin, Jia-Ren, Benjamin Izar, Shu Wang, Clarence Yapp, Shaolin Mei, Parin M. Shah, Sandro Santagata, and Peter K. Sorger. 2018. “Highly Multiplexed Immunofluorescence Imaging of Human Tissues and Tumors Using T-Cycif and Conventional Optical Microscopes.” <em>eLife</em> 7: 1–46.</p>
</div>
<div id="ref-Lubeck2014">
<p>Lubeck, Eric, Ahmet F Coskun, Timur Zhiyentayev, Mubhij Ahmad, and Long Cai. 2014. “Single-Cell in Situ Rna Profiling by Sequential Hybridization.” <em>Nature Methods</em> 11: 360–61.</p>
</div>
<div id="ref-Shapiro2017">
<p>Schapiro, Denis, Hartland W Jackson, Swetha Raghuraman, Jana R Fischer, Vito RT Zanotelli, Daniel Schulz, Charlotte Giesen, Raúl Catena, Zsuzsanna Varga, and Bernd Bodenmiller. 2017. “HistoCAT: Analysis of Cell Phenotypes and Interactions in Multiplex Image Cytometry Data.” <em>Nature Methods</em> 14: 873–76.</p>
</div>
<div id="ref-Schulz2018">
<p>Schulz, Daniel, Vito RT Zanotelli, Rana R Fischer, Denis Schapiro, Stefanie Engler, Xiao-Kang Lun, Hartland W Jackson, and Bernd Bodenmiller. 2018. “Simultaneous Multiplexed Imaging of mRNA and Proteins with Subcellular Resolution in Breast Cancer Tissue Samples by Mass Cytometry.” <em>Cell Systems</em> 6: 25–36.e5.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Nils Eling.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
