<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tools for IMC data analysis • imcRtools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tools for IMC data analysis">
<meta property="og:description" content="imcRtools">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">imcRtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/imcRtools.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BodenmillerGroup/imcRtools/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="imcRtools_files/htmlwidgets-1.6.2/htmlwidgets.js"></script><link href="imcRtools_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="imcRtools_files/datatables-binding-0.30/datatables.js"></script><link href="imcRtools_files/dt-core-1.13.4/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="imcRtools_files/dt-core-1.13.4/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="imcRtools_files/dt-core-1.13.4/js/jquery.dataTables.min.js"></script><link href="imcRtools_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="imcRtools_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tools for IMC data analysis</h1>
                        <h4 data-toc-skip class="author">Nils Eling</h4>
            <address class="author_afil">
      Department for Quantitative Biomedicine, University of
ZurichInstitute for Molecular Health Sciences, ETH
Zurich<br><a class="author_email" href="mailto:#"></a><a href="mailto:nils.eling@dqbm.uzh.ch" class="email">nils.eling@dqbm.uzh.ch</a>
      </address>
                              <h4 data-toc-skip class="author">Lasse
Meyer</h4>
            <address class="author_afil">
      Department for Quantitative Biomedicine, University of
ZurichInstitute for Molecular Health Sciences, ETH
Zurich<br><a class="author_email" href="mailto:#">lasse.meyer@<span class="citation">(<strong>uzh.ch?</strong>)</span></a>
      </address>
                              <h4 data-toc-skip class="author">Daniel
Schulz</h4>
            <address class="author_afil">
      Department for Quantitative Biomedicine, University of
ZurichInstitute for Molecular Health Sciences, ETH
Zurich<br><a class="author_email" href="mailto:#">daniel.schulz@<span class="citation">(<strong>uzh.ch?</strong>)</span></a>
      </address>
                  
            <h4 data-toc-skip class="date">20 November 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BodenmillerGroup/imcRtools/blob/HEAD/vignettes/imcRtools.Rmd" class="external-link"><code>vignettes/imcRtools.Rmd</code></a></small>
      <div class="hidden name"><code>imcRtools.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This R package supports the handling and analysis of imaging
      mass cytometry and other highly multiplexed imaging data. The main
      functionality includes reading in single-cell data after image
      segmentation and measurement, data formatting to perform channel
      spillover correction and a number of spatial analysis approaches.
      First, cell-cell interactions are detected via spatial graph
      construction; these graphs can be visualized with cells
      representing nodes and interactions representing edges.
      Furthermore, per cell, its direct neighbours are summarized to
      allow spatial clustering. Per image/grouping level, interactions
      between types of cells are counted, averaged and compared against
      random permutations. In that way, types of cells that interact
      more (attraction) or less (avoidance) frequently than expected by
      chance are detected.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette gives an introduction to handling and analyzing imaging
mass cytometry (IMC) and other highly-multiplexed imaging data in R. The
<code>imcRtools</code> package relies on expression and morphological
features extracted from multi-channel images using corresponding
segmentation masks. A description of data types and segmentation
approaches can be found below (<a href="#dataTypes">data types</a>, <a href="#segmentation">segmentation</a>). However, due to shared data
structures, the functionalities of the <code>imcRtools</code> package
are applicable to most highly multiplexed imaging modalities.</p>
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>The <code>imcRtools</code> package exports functions and example data
to perform the following analyses:</p>
<ol style="list-style-type: decimal">
<li><a href="#read-in-data">Read in pre-processed data</a></li>
<li><a href="#spillover">Perform spillover correction for IMC
data</a></li>
<li>
<a href="#build-graph">Build</a> and <a href="#viz-cells">visualize</a> spatial graphs</li>
<li>
<a href="#aggregate-neighbors">Aggregate across neighbouring
cells</a> for spatial clustering</li>
<li>
<a href="#patch-detection">Detect spatial patches</a> of similar
cell-types</li>
<li><a href="#test-neighborhood">Test the attraction or avoidance
between celltypes</a></li>
</ol>
<p>To highlight the usability of these function, the
<code>imcRtools</code> package also exports a number of <a href="#test-data">test data files</a>.</p>
</div>
<div class="section level3">
<h3 id="highly-multiplexed-imaging">Highly multiplexed imaging<a class="anchor" aria-label="anchor" href="#highly-multiplexed-imaging"></a>
</h3>
<p>Highly multiplexed imaging techniques <span class="citation">(Bodenmiller 2016)</span> such as imaging mass
cytometry (IMC) <span class="citation">(Giesen et al. 2014)</span>,
multiplexed ion beam imaging (MIBI) <span class="citation">(Angelo et
al. 2014)</span> and cyclic immunofluorescence techniques <span class="citation">Gut, Herrmann, and Pelkmans (2018)</span> acquire
read-outs of the expression of tens of protein in a spatially resolved
manner. After technology-dependent data pre-processing, the raw data
files are comparable: multi-channel images for the dimensions
<code>x</code>, <code>y</code>, and <code>c</code>, where <code>x</code>
and <code>y</code> define the number of pixels (<code>x * y</code>) per
image and <code>c</code> the number of proteins (also refered to as
“markers”) measured per image. More info on the <a href="#dataTypes">data types</a> and <a href="#segmentation">further
pre-processing</a> can be found below.</p>
<p>Increased multiplexity compared to epitope-based techniques is
achieved using single-cell resolved spatial transcriptomics techniques
including MERFISH <span class="citation">(Chen et al. 2015)</span> and
seqFISH <span class="citation">(Lubeck et al. 2014)</span>. However,
data produced by these techniques required different pre-processing
steps. Nevertheless, analysis performed on the single-cell level is
equally applicable.</p>
<div class="section level4">
<h4 id="imaging-mass-cytometry">Imaging mass cytometry<a class="anchor" aria-label="anchor" href="#imaging-mass-cytometry"></a>
</h4>
<p>IMC <span class="citation">(Giesen et al. 2014)</span> is a highly
multiplexed imaging approach to measure spatial protein abundance. In
IMC, tissue sections on slides are stained with a mix of around 40
metal-conjugated antibodies prior to laser ablation with <span class="math inline">\(1\mu{}m\)</span> resolution. The ablated material
is transferred to a mass cytometer for time-of-flight detection of the
metal ions <span class="citation">(Giesen et al. 2014)</span>. At an
ablation frequency of 200Hz, a 1mm x 1mm region can be acquired within
1.5 hours.</p>
</div>
<div class="section level4">
<h4 id="dataTypes">Data types<a class="anchor" aria-label="anchor" href="#dataTypes"></a>
</h4>
<p>In the case of IMC, the raw data output are .mcd files containing all
acquired regions per slide. During image pre-prcoessing these files are
converted into individual multi-channel .tiff and <a href="https://docs.openmicroscopy.org/ome-model/5.6.3/ome-tiff/" class="external-link">OME-TIFF</a>
files. These file formats are supported by most open-source and
commercial image analysis software, such as <a href="https://imagej.net/software/fiji/" class="external-link">Fiji</a>, <a href="https://qupath.github.io/" class="external-link">QuPath</a> or <a href="https://napari.org/" class="external-link">napari</a>.</p>
<p>In R, these .tiff files can be read into individual <a href="https://bioconductor.org/packages/release/bioc/vignettes/EBImage/inst/doc/EBImage-introduction.html#3_Image_data_representation" class="external-link">Image</a>
objects or combined into a <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/cytomapper/inst/doc/cytomapper.html#5_The_CytoImageList_object" class="external-link">CytoImageList</a>
object supported by the <a href="https://www.bioconductor.org/packages/release/bioc/html/cytomapper.html" class="external-link">cytomapper</a>
package.</p>
</div>
<div class="section level4">
<h4 id="segmentation">Segmentation and feature extraction<a class="anchor" aria-label="anchor" href="#segmentation"></a>
</h4>
<p>The pixel resolution of most highly multiplexed imaging technologies
including IMC support the detection of cellular structures. Therefore, a
common processing step using multi-channel images is object
segmentation. In this vignette, objects are defined as cells; however,
also larger scale structures could be segmented.</p>
<p>There are multiple different image segmentation approaches available.
However, <code>imcRtools</code> only supports direct reader functions
for two segmentation strategies developed for highly multiplexed imaging
technologies:</p>
<ol style="list-style-type: decimal">
<li><p>The <a href="https://github.com/BodenmillerGroup/ImcSegmentationPipeline" class="external-link">ImcSegmentationPipeline</a>
has been developed to give the user flexibility in how to perform
channel selection and image segmentation. The underlying principle is to
train a pixel classifier (using <a href="https://www.ilastik.org/" class="external-link">ilastik</a>) on a number of selected
channels to perform probabilistic classification of each pixel into:
background, cytoplasm and nucleus. Based on these classification
probabilities a <a href="https://cellprofiler.org/" class="external-link">CellProfiler</a>
pipeline performs cell segmentation and feature extraction.</p></li>
<li><p>A containerized version of this pipeline is implemented in <a href="https://github.com/BodenmillerGroup/steinbock" class="external-link">steinbock</a>.
<code>steinbock</code> further supports segmentation via the use of
<code>Mesmer</code> from the <code>DeepCell</code> library <span class="citation">(Greenwald et al. 2021)</span>.</p></li>
</ol>
<p>While the output of both approaches is structured differently, the
exported features are comparable:</p>
<ol style="list-style-type: decimal">
<li>per cell: channel intensity, morphology and location</li>
<li>cell-cell interactions exported as graph</li>
</ol>
<p>By combining these with availabel channel information, the data can
be read into a <em><a href="https://bioconductor.org/packages/3.18/SpatialExperiment" class="external-link">SpatialExperiment</a></em>
or <em><a href="https://bioconductor.org/packages/3.18/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
object (see <a href="#read-in-data">below</a>).</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="test-data">Example data<a class="anchor" aria-label="anchor" href="#test-data"></a>
</h2>
<p>The <code>imcRtools</code> package contains a number of example data
generated by the Hyperion imaging system for different purposes. The
following section gives an overview of these files.</p>
<div class="section level3">
<h3 id="for-spillover-correction">For spillover correction<a class="anchor" aria-label="anchor" href="#for-spillover-correction"></a>
</h3>
<p>To highlight the use of the <code>imcRtools</code> package for
spillover correction, we provide four .txt files containing pixel
intensities of four spotted metals.</p>
<p>Please refer to the <a href="#spillover">spillover correction</a>
section for full details.</p>
<p>These files are accessible via:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/spillover"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">path</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Dy161.txt" "Dy162.txt" "Dy163.txt" "Dy164.txt"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="raw-data-in-form-of--txt-files">Raw data in form of .txt files<a class="anchor" aria-label="anchor" href="#raw-data-in-form-of--txt-files"></a>
</h3>
<p>IMC generates .mcd files storing the raw data or all acquired regions
of interest (ROI). In addition, the raw pixel values are also stored in
individual .txt files per ROI.</p>
<p>To highlight reading in raw data in form of .txt files, the
<code>imcRtools</code> contains 3 sample acquisitions:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">txt_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mockData/raw"</span>, </span>
<span>                                    package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">txt_files</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "20210305_NE_mockData2_ROI_001_1.txt" "20210305_NE_mockData2_ROI_002_2.txt"</span></span>
<span><span class="co">## [3] "20210305_NE_mockData2_ROI_003_3.txt"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="imcsegmentationpipeline-output-data">ImcSegmentationPipeline output data<a class="anchor" aria-label="anchor" href="#imcsegmentationpipeline-output-data"></a>
</h3>
<p>IMC data preprocessing and segmentation can be performed using the <a href="https://github.com/BodenmillerGroup/ImcSegmentationPipeline" class="external-link">ImcSegmentationPipeline</a>.
It generates a number of <code>.csv</code> files containing
object/cell-specific and image-specific metadata.</p>
<p>The <code>imcRtools</code> package exports the
<code>read_cpout</code> function as convenient reader function for
outputs generated by the <code>ImcSegmentationPipeline</code>. For
demonstration purposes, <code>imcRtools</code> contains the output of
running the pipeline on a small example dataset:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mockData/cpout"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">path</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "cell.csv"                 "Experiment.csv"          </span></span>
<span><span class="co">## [3] "Image.csv"                "Object_relationships.csv"</span></span>
<span><span class="co">## [5] "panel.csv"                "var_cell.csv"            </span></span>
<span><span class="co">## [7] "var_Image.csv"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="steinbock-output-data">steinbock output data<a class="anchor" aria-label="anchor" href="#steinbock-output-data"></a>
</h3>
<p>The <a href="https://github.com/BodenmillerGroup/steinbock" class="external-link">steinbock</a>
pipeline can be used to process, segment and extract features from IMC
data. For more information, please refer to its <a href="https://bodenmillergroup.github.io/steinbock/" class="external-link">documentation</a>.</p>
<p>To highlight the functionality of <code>imcRtools</code> to read in
single-cell data generated by <code>steinbock</code>, we provide a small
toy dataset available at:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mockData/steinbock"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">path</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "images.csv"                               </span></span>
<span><span class="co">##  [2] "intensities/20210305_NE_mockData1_001.csv"</span></span>
<span><span class="co">##  [3] "intensities/20210305_NE_mockData1_002.csv"</span></span>
<span><span class="co">##  [4] "intensities/20210305_NE_mockData1_003.csv"</span></span>
<span><span class="co">##  [5] "intensities/20210305_NE_mockData2_001.csv"</span></span>
<span><span class="co">##  [6] "intensities/20210305_NE_mockData2_002.csv"</span></span>
<span><span class="co">##  [7] "intensities/20210305_NE_mockData2_003.csv"</span></span>
<span><span class="co">##  [8] "intensities/20210305_NE_mockData3_001.csv"</span></span>
<span><span class="co">##  [9] "intensities/20210305_NE_mockData3_002.csv"</span></span>
<span><span class="co">## [10] "intensities/20210305_NE_mockData3_003.csv"</span></span>
<span><span class="co">## [11] "intensities/20210305_NE_mockData4_001.csv"</span></span>
<span><span class="co">## [12] "intensities/20210305_NE_mockData4_002.csv"</span></span>
<span><span class="co">## [13] "intensities/20210305_NE_mockData4_003.csv"</span></span>
<span><span class="co">## [14] "intensities/20210305_NE_mockData5_001.csv"</span></span>
<span><span class="co">## [15] "intensities/20210305_NE_mockData5_002.csv"</span></span>
<span><span class="co">## [16] "intensities/20210305_NE_mockData5_003.csv"</span></span>
<span><span class="co">## [17] "neighbors/20210305_NE_mockData1_001.csv"  </span></span>
<span><span class="co">## [18] "neighbors/20210305_NE_mockData1_002.csv"  </span></span>
<span><span class="co">## [19] "neighbors/20210305_NE_mockData1_003.csv"  </span></span>
<span><span class="co">## [20] "neighbors/20210305_NE_mockData2_001.csv"  </span></span>
<span><span class="co">## [21] "neighbors/20210305_NE_mockData2_002.csv"  </span></span>
<span><span class="co">## [22] "neighbors/20210305_NE_mockData2_003.csv"  </span></span>
<span><span class="co">## [23] "neighbors/20210305_NE_mockData3_001.csv"  </span></span>
<span><span class="co">## [24] "neighbors/20210305_NE_mockData3_002.csv"  </span></span>
<span><span class="co">## [25] "neighbors/20210305_NE_mockData3_003.csv"  </span></span>
<span><span class="co">## [26] "neighbors/20210305_NE_mockData4_001.csv"  </span></span>
<span><span class="co">## [27] "neighbors/20210305_NE_mockData4_002.csv"  </span></span>
<span><span class="co">## [28] "neighbors/20210305_NE_mockData4_003.csv"  </span></span>
<span><span class="co">## [29] "neighbors/20210305_NE_mockData5_001.csv"  </span></span>
<span><span class="co">## [30] "neighbors/20210305_NE_mockData5_002.csv"  </span></span>
<span><span class="co">## [31] "neighbors/20210305_NE_mockData5_003.csv"  </span></span>
<span><span class="co">## [32] "panel.csv"                                </span></span>
<span><span class="co">## [33] "regionprops/20210305_NE_mockData1_001.csv"</span></span>
<span><span class="co">## [34] "regionprops/20210305_NE_mockData1_002.csv"</span></span>
<span><span class="co">## [35] "regionprops/20210305_NE_mockData1_003.csv"</span></span>
<span><span class="co">## [36] "regionprops/20210305_NE_mockData2_001.csv"</span></span>
<span><span class="co">## [37] "regionprops/20210305_NE_mockData2_002.csv"</span></span>
<span><span class="co">## [38] "regionprops/20210305_NE_mockData2_003.csv"</span></span>
<span><span class="co">## [39] "regionprops/20210305_NE_mockData3_001.csv"</span></span>
<span><span class="co">## [40] "regionprops/20210305_NE_mockData3_002.csv"</span></span>
<span><span class="co">## [41] "regionprops/20210305_NE_mockData3_003.csv"</span></span>
<span><span class="co">## [42] "regionprops/20210305_NE_mockData4_001.csv"</span></span>
<span><span class="co">## [43] "regionprops/20210305_NE_mockData4_002.csv"</span></span>
<span><span class="co">## [44] "regionprops/20210305_NE_mockData4_003.csv"</span></span>
<span><span class="co">## [45] "regionprops/20210305_NE_mockData5_001.csv"</span></span>
<span><span class="co">## [46] "regionprops/20210305_NE_mockData5_002.csv"</span></span>
<span><span class="co">## [47] "regionprops/20210305_NE_mockData5_003.csv"</span></span>
<span><span class="co">## [48] "steinbock.sh"</span></span></code></pre>
<p>The example data related to the <code>ImcSegmentationPipeline</code>
and <code>steinbock</code> can also be accessed <a href="https://github.com/BodenmillerGroup/TestData/tree/main/datasets/210308_ImcTestData" class="external-link">online</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="read-in-data">Read in IMC data<a class="anchor" aria-label="anchor" href="#read-in-data"></a>
</h2>
<p>The <code>imcRtools</code> package supports reading in data generated
by the <a href="https://github.com/BodenmillerGroup/ImcSegmentationPipeline" class="external-link">ImcSegmentationPipeline</a>
or <a href="https://github.com/BodenmillerGroup/steinbock" class="external-link">steinbock</a>
pipeline.</p>
<p>To read in the outpout data into a <em><a href="https://bioconductor.org/packages/3.18/SpatialExperiment" class="external-link">SpatialExperiment</a></em>
or <em><a href="https://bioconductor.org/packages/3.18/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>,
the <code>imcRtools</code> package exports the <code>read_cpout</code>
function.</p>
<p>By default, the single-cell data is read into a <em><a href="https://bioconductor.org/packages/3.18/SpatialExperiment" class="external-link">SpatialExperiment</a></em>
object. Here, the extracted channel- and cell-specific intensities are
stored in the <code>counts(spe)</code> slot. All morphological features
are stored in <code>colData(spe)</code> and the spatial locations of the
cells are stored in <code>spatialCoords(spe)</code>. The interaction
graph is stored in <code>colPair(spe, "neighbourhood")</code>.</p>
<p>Alternatively, the data can be read into a <em><a href="https://bioconductor.org/packages/3.18/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
object. The only difference is the lack of
<code>spatialCoords(sce)</code>. Here, the spatial coordinates are
stored in <code>colData(spe)$Pos_X</code> and
<code>colData(spe)$Pos_Y</code>.</p>
<div class="section level3">
<h3 id="read-in-cellprofiler-output">Read in CellProfiler output<a class="anchor" aria-label="anchor" href="#read-in-cellprofiler-output"></a>
</h3>
<p>The <a href="https://github.com/BodenmillerGroup/ImcSegmentationPipeline" class="external-link">ImcSegmentationPipeline</a>
produces a number of output files. By default, all single-cell features
are measured and exported. To browse and select the features of
interest, the <code>imcRtools</code> package provides the
<code>show_cpout_features</code> function:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mockData/cpout"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/show_cpout_features.html">show_cpout_features</a></span><span class="op">(</span><span class="va">path</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-bcbf098a310ec780179d" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-bcbf098a310ec780179d">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200","201","202","203","204"],["Location_Center_X","Location_Center_Y","Location_Center_Z","Number_Object_Number","Neighbors_NumberOfNeighbors_8","Neighbors_PercentTouching_8","Neighbors_FirstClosestObjectNumber_8","Neighbors_FirstClosestDistance_8","Neighbors_SecondClosestObjectNumber_8","Neighbors_SecondClosestDistance_8","Neighbors_AngleBetweenNeighbors_8","Intensity_IntegratedIntensity_FullStack_c1","Intensity_MeanIntensity_FullStack_c1","Intensity_StdIntensity_FullStack_c1","Intensity_MinIntensity_FullStack_c1","Intensity_MaxIntensity_FullStack_c1","Intensity_IntegratedIntensityEdge_FullStack_c1","Intensity_MeanIntensityEdge_FullStack_c1","Intensity_StdIntensityEdge_FullStack_c1","Intensity_MinIntensityEdge_FullStack_c1","Intensity_MaxIntensityEdge_FullStack_c1","Intensity_MassDisplacement_FullStack_c1","Intensity_LowerQuartileIntensity_FullStack_c1","Intensity_MedianIntensity_FullStack_c1","Intensity_MADIntensity_FullStack_c1","Intensity_UpperQuartileIntensity_FullStack_c1","Location_CenterMassIntensity_X_FullStack_c1","Location_CenterMassIntensity_Y_FullStack_c1","Location_CenterMassIntensity_Z_FullStack_c1","Location_MaxIntensity_X_FullStack_c1","Location_MaxIntensity_Y_FullStack_c1","Location_MaxIntensity_Z_FullStack_c1","Intensity_IntegratedIntensity_FullStack_c2","Intensity_MeanIntensity_FullStack_c2","Intensity_StdIntensity_FullStack_c2","Intensity_MinIntensity_FullStack_c2","Intensity_MaxIntensity_FullStack_c2","Intensity_IntegratedIntensityEdge_FullStack_c2","Intensity_MeanIntensityEdge_FullStack_c2","Intensity_StdIntensityEdge_FullStack_c2","Intensity_MinIntensityEdge_FullStack_c2","Intensity_MaxIntensityEdge_FullStack_c2","Intensity_MassDisplacement_FullStack_c2","Intensity_LowerQuartileIntensity_FullStack_c2","Intensity_MedianIntensity_FullStack_c2","Intensity_MADIntensity_FullStack_c2","Intensity_UpperQuartileIntensity_FullStack_c2","Location_CenterMassIntensity_X_FullStack_c2","Location_CenterMassIntensity_Y_FullStack_c2","Location_CenterMassIntensity_Z_FullStack_c2","Location_MaxIntensity_X_FullStack_c2","Location_MaxIntensity_Y_FullStack_c2","Location_MaxIntensity_Z_FullStack_c2","Intensity_IntegratedIntensity_FullStack_c3","Intensity_MeanIntensity_FullStack_c3","Intensity_StdIntensity_FullStack_c3","Intensity_MinIntensity_FullStack_c3","Intensity_MaxIntensity_FullStack_c3","Intensity_IntegratedIntensityEdge_FullStack_c3","Intensity_MeanIntensityEdge_FullStack_c3","Intensity_StdIntensityEdge_FullStack_c3","Intensity_MinIntensityEdge_FullStack_c3","Intensity_MaxIntensityEdge_FullStack_c3","Intensity_MassDisplacement_FullStack_c3","Intensity_LowerQuartileIntensity_FullStack_c3","Intensity_MedianIntensity_FullStack_c3","Intensity_MADIntensity_FullStack_c3","Intensity_UpperQuartileIntensity_FullStack_c3","Location_CenterMassIntensity_X_FullStack_c3","Location_CenterMassIntensity_Y_FullStack_c3","Location_CenterMassIntensity_Z_FullStack_c3","Location_MaxIntensity_X_FullStack_c3","Location_MaxIntensity_Y_FullStack_c3","Location_MaxIntensity_Z_FullStack_c3","Intensity_IntegratedIntensity_FullStack_c4","Intensity_MeanIntensity_FullStack_c4","Intensity_StdIntensity_FullStack_c4","Intensity_MinIntensity_FullStack_c4","Intensity_MaxIntensity_FullStack_c4","Intensity_IntegratedIntensityEdge_FullStack_c4","Intensity_MeanIntensityEdge_FullStack_c4","Intensity_StdIntensityEdge_FullStack_c4","Intensity_MinIntensityEdge_FullStack_c4","Intensity_MaxIntensityEdge_FullStack_c4","Intensity_MassDisplacement_FullStack_c4","Intensity_LowerQuartileIntensity_FullStack_c4","Intensity_MedianIntensity_FullStack_c4","Intensity_MADIntensity_FullStack_c4","Intensity_UpperQuartileIntensity_FullStack_c4","Location_CenterMassIntensity_X_FullStack_c4","Location_CenterMassIntensity_Y_FullStack_c4","Location_CenterMassIntensity_Z_FullStack_c4","Location_MaxIntensity_X_FullStack_c4","Location_MaxIntensity_Y_FullStack_c4","Location_MaxIntensity_Z_FullStack_c4","Intensity_IntegratedIntensity_FullStack_c5","Intensity_MeanIntensity_FullStack_c5","Intensity_StdIntensity_FullStack_c5","Intensity_MinIntensity_FullStack_c5","Intensity_MaxIntensity_FullStack_c5","Intensity_IntegratedIntensityEdge_FullStack_c5","Intensity_MeanIntensityEdge_FullStack_c5","Intensity_StdIntensityEdge_FullStack_c5","Intensity_MinIntensityEdge_FullStack_c5","Intensity_MaxIntensityEdge_FullStack_c5","Intensity_MassDisplacement_FullStack_c5","Intensity_LowerQuartileIntensity_FullStack_c5","Intensity_MedianIntensity_FullStack_c5","Intensity_MADIntensity_FullStack_c5","Intensity_UpperQuartileIntensity_FullStack_c5","Location_CenterMassIntensity_X_FullStack_c5","Location_CenterMassIntensity_Y_FullStack_c5","Location_CenterMassIntensity_Z_FullStack_c5","Location_MaxIntensity_X_FullStack_c5","Location_MaxIntensity_Y_FullStack_c5","Location_MaxIntensity_Z_FullStack_c5","Intensity_IntegratedIntensity_ProbabStack_c1","Intensity_MeanIntensity_ProbabStack_c1","Intensity_StdIntensity_ProbabStack_c1","Intensity_MinIntensity_ProbabStack_c1","Intensity_MaxIntensity_ProbabStack_c1","Intensity_IntegratedIntensityEdge_ProbabStack_c1","Intensity_MeanIntensityEdge_ProbabStack_c1","Intensity_StdIntensityEdge_ProbabStack_c1","Intensity_MinIntensityEdge_ProbabStack_c1","Intensity_MaxIntensityEdge_ProbabStack_c1","Intensity_MassDisplacement_ProbabStack_c1","Intensity_LowerQuartileIntensity_ProbabStack_c1","Intensity_MedianIntensity_ProbabStack_c1","Intensity_MADIntensity_ProbabStack_c1","Intensity_UpperQuartileIntensity_ProbabStack_c1","Location_CenterMassIntensity_X_ProbabStack_c1","Location_CenterMassIntensity_Y_ProbabStack_c1","Location_CenterMassIntensity_Z_ProbabStack_c1","Location_MaxIntensity_X_ProbabStack_c1","Location_MaxIntensity_Y_ProbabStack_c1","Location_MaxIntensity_Z_ProbabStack_c1","Intensity_IntegratedIntensity_ProbabStack_c2","Intensity_MeanIntensity_ProbabStack_c2","Intensity_StdIntensity_ProbabStack_c2","Intensity_MinIntensity_ProbabStack_c2","Intensity_MaxIntensity_ProbabStack_c2","Intensity_IntegratedIntensityEdge_ProbabStack_c2","Intensity_MeanIntensityEdge_ProbabStack_c2","Intensity_StdIntensityEdge_ProbabStack_c2","Intensity_MinIntensityEdge_ProbabStack_c2","Intensity_MaxIntensityEdge_ProbabStack_c2","Intensity_MassDisplacement_ProbabStack_c2","Intensity_LowerQuartileIntensity_ProbabStack_c2","Intensity_MedianIntensity_ProbabStack_c2","Intensity_MADIntensity_ProbabStack_c2","Intensity_UpperQuartileIntensity_ProbabStack_c2","Location_CenterMassIntensity_X_ProbabStack_c2","Location_CenterMassIntensity_Y_ProbabStack_c2","Location_CenterMassIntensity_Z_ProbabStack_c2","Location_MaxIntensity_X_ProbabStack_c2","Location_MaxIntensity_Y_ProbabStack_c2","Location_MaxIntensity_Z_ProbabStack_c2","Intensity_IntegratedIntensity_ProbabStack_c3","Intensity_MeanIntensity_ProbabStack_c3","Intensity_StdIntensity_ProbabStack_c3","Intensity_MinIntensity_ProbabStack_c3","Intensity_MaxIntensity_ProbabStack_c3","Intensity_IntegratedIntensityEdge_ProbabStack_c3","Intensity_MeanIntensityEdge_ProbabStack_c3","Intensity_StdIntensityEdge_ProbabStack_c3","Intensity_MinIntensityEdge_ProbabStack_c3","Intensity_MaxIntensityEdge_ProbabStack_c3","Intensity_MassDisplacement_ProbabStack_c3","Intensity_LowerQuartileIntensity_ProbabStack_c3","Intensity_MedianIntensity_ProbabStack_c3","Intensity_MADIntensity_ProbabStack_c3","Intensity_UpperQuartileIntensity_ProbabStack_c3","Location_CenterMassIntensity_X_ProbabStack_c3","Location_CenterMassIntensity_Y_ProbabStack_c3","Location_CenterMassIntensity_Z_ProbabStack_c3","Location_MaxIntensity_X_ProbabStack_c3","Location_MaxIntensity_Y_ProbabStack_c3","Location_MaxIntensity_Z_ProbabStack_c3","AreaShape_Extent","AreaShape_EulerNumber","AreaShape_EquivalentDiameter","AreaShape_MajorAxisLength","AreaShape_MinorAxisLength","AreaShape_Center_X","AreaShape_Center_Y","AreaShape_BoundingBoxMinimum_X","AreaShape_BoundingBoxMinimum_Y","AreaShape_BoundingBoxMaximum_X","AreaShape_BoundingBoxMaximum_Y","AreaShape_Area","AreaShape_Perimeter","AreaShape_MaximumRadius","AreaShape_MeanRadius","AreaShape_MedianRadius","AreaShape_MinFeretDiameter","AreaShape_MaxFeretDiameter","AreaShape_Orientation","AreaShape_Eccentricity","AreaShape_FormFactor","AreaShape_Solidity","AreaShape_ConvexArea","AreaShape_Compactness","AreaShape_BoundingBoxArea"],["Location","Location","Location","ID","Neighbors","Neighbors","Neighbors","Neighbors","Neighbors","Neighbors","Neighbors","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Intensity","Location","Location","Location","Location","Location","Location","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape","AreaShape"],[null,null,null,null,null,null,null,null,null,null,null,"FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","FullStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack","ProbabStack",null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],["Center_X","Center_Y","Center_Z","Number_Object_Number","NumberOfNeighbors","PercentTouching","FirstClosestObjectNumber","FirstClosestDistance","SecondClosestObjectNumber","SecondClosestDistance","AngleBetweenNeighbors","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","IntegratedIntensity","MeanIntensity","StdIntensity","MinIntensity","MaxIntensity","IntegratedIntensityEdge","MeanIntensityEdge","StdIntensityEdge","MinIntensityEdge","MaxIntensityEdge","MassDisplacement","LowerQuartileIntensity","MedianIntensity","MADIntensity","UpperQuartileIntensity","CenterMassIntensity_X","CenterMassIntensity_Y","CenterMassIntensity_Z","MaxIntensity_X","MaxIntensity_Y","MaxIntensity_Z","Extent","EulerNumber","EquivalentDiameter","MajorAxisLength","MinorAxisLength","Center_X","Center_Y","BoundingBoxMinimum_X","BoundingBoxMinimum_Y","BoundingBoxMaximum_X","BoundingBoxMaximum_Y","Area","Perimeter","MaximumRadius","MeanRadius","MedianRadius","MinFeretDiameter","MaxFeretDiameter","Orientation","Eccentricity","FormFactor","Solidity","ConvexArea","Compactness","BoundingBoxArea"],[null,null,null,null,null,null,null,null,null,null,null,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],[null,null,null,null,8,8,8,8,8,8,8,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],[null,null,null,null,null,null,null,null,null,null,null,"Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Ag107","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Pr141","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Sm147","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Eu153","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","Yb172","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellCenter","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","CellBorder","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background","Background",null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],["float","float","float","integer","integer","float","integer","float","integer","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float","float"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>column_name<\/th>\n      <th>category<\/th>\n      <th>image_name<\/th>\n      <th>object_name<\/th>\n      <th>feature_name<\/th>\n      <th>channel<\/th>\n      <th>parameters<\/th>\n      <th>channel_id<\/th>\n      <th>data_type<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[6,7]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script><p>By default, <code>read_cpout</code> will read in the mean intensity
per channel and cell from “hot pixel” filtered image stacks specified
via
<code>intensities = "Intensity_MeanIntensity_FullStackFiltered"</code>.
Please refer to <code><a href="../reference/read_cpout.html">?read_cpout</a></code> for the full documentation.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cur_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mockData/cpout"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Read as SpatialExperiment</span></span>
<span><span class="op">(</span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_cpout.html">read_cpout</a></span><span class="op">(</span><span class="va">cur_path</span>, graph_file <span class="op">=</span> <span class="st">"Object_relationships.csv"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SpatialExperiment </span></span>
<span><span class="co">## dim: 5 239 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(5): Ag107 Pr141 Sm147 Eu153 Yb172</span></span>
<span><span class="co">## rowData names(7): Tube.Number Metal.Tag ... deepcell cellpose</span></span>
<span><span class="co">## colnames: NULL</span></span>
<span><span class="co">## colData names(12): sample_id ObjectNumber ... Metadata_acid</span></span>
<span><span class="co">##   Metadata_description</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## spatialCoords names(2) : Pos_X Pos_Y</span></span>
<span><span class="co">## imgData names(1): sample_id</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read as SingleCellExperiment</span></span>
<span><span class="op">(</span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_cpout.html">read_cpout</a></span><span class="op">(</span><span class="va">cur_path</span>, graph_file <span class="op">=</span> <span class="st">"Object_relationships.csv"</span>, </span>
<span>                   return_as <span class="op">=</span> <span class="st">"sce"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 5 239 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(5): Ag107 Pr141 Sm147 Eu153 Yb172</span></span>
<span><span class="co">## rowData names(7): Tube.Number Metal.Tag ... deepcell cellpose</span></span>
<span><span class="co">## colnames: NULL</span></span>
<span><span class="co">## colData names(14): sample_id ObjectNumber ... Metadata_acid</span></span>
<span><span class="co">##   Metadata_description</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="read-in-steinbock-output">Read in steinbock output<a class="anchor" aria-label="anchor" href="#read-in-steinbock-output"></a>
</h3>
<p>Single-cell data and all associated metadata (e.g. spatial location,
morphology and interaction graphs) as produced by the <a href="https://github.com/BodenmillerGroup/steinbock" class="external-link">steinbock</a>
pipeline can be read in using the <code>read_steinbock</code>
function:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cur_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mockData/steinbock"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Read as SpatialExperiment</span></span>
<span><span class="op">(</span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_steinbock.html">read_steinbock</a></span><span class="op">(</span><span class="va">cur_path</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SpatialExperiment </span></span>
<span><span class="co">## dim: 5 404 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(5): Ag107 Cytokeratin 5 Laminin YBX1 H3K27Ac</span></span>
<span><span class="co">## rowData names(7): channel name ... cellpose Tube.Number</span></span>
<span><span class="co">## colnames: NULL</span></span>
<span><span class="co">## colData names(8): sample_id ObjectNumber ... width_px height_px</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## spatialCoords names(2) : Pos_X Pos_Y</span></span>
<span><span class="co">## imgData names(1): sample_id</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read as SingleCellExperiment</span></span>
<span><span class="op">(</span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_steinbock.html">read_steinbock</a></span><span class="op">(</span><span class="va">cur_path</span>, return_as <span class="op">=</span> <span class="st">"sce"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 5 404 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(5): Ag107 Cytokeratin 5 Laminin YBX1 H3K27Ac</span></span>
<span><span class="co">## rowData names(7): channel name ... cellpose Tube.Number</span></span>
<span><span class="co">## colnames: NULL</span></span>
<span><span class="co">## colData names(10): sample_id ObjectNumber ... width_px height_px</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>For more information, please refer to
<code><a href="../reference/read_steinbock.html">?read_steinbock</a></code>.</p>
</div>
<div class="section level3">
<h3 id="read-raw--txt-files-into-image-objects">Read raw .txt files into Image objects<a class="anchor" aria-label="anchor" href="#read-raw--txt-files-into-image-objects"></a>
</h3>
<p>For reading in and visualization of multi-channel images and
segmentation masks, please refer to the <em><a href="https://bioconductor.org/packages/3.18/cytomapper" class="external-link">cytomapper</a></em>
package. The <code>imcRtools</code> package however supports reading in
raw .txt files generated by the Hyperion imaging system into a
<code>CytoImageList</code> object; a data container exported by
<code>cytomapper</code>.</p>
<p>The user needs to provide a path from which all .txt files will be
read in:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mockData/raw"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cur_CytoImageList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readImagefromTXT.html">readImagefromTXT</a></span><span class="op">(</span><span class="va">path</span><span class="op">)</span></span>
<span><span class="va">cur_CytoImageList</span></span></code></pre></div>
<pre><code><span><span class="co">## CytoImageList containing 3 image(s)</span></span>
<span><span class="co">## names(3): 20210305_NE_mockData2_ROI_001_1 20210305_NE_mockData2_ROI_002_2 20210305_NE_mockData2_ROI_003_3 </span></span>
<span><span class="co">## Each image contains 5 channel(s)</span></span>
<span><span class="co">## channelNames(5): Ag107Di Pr141Di Sm147Di Eu153Di Yb172Di</span></span></code></pre>
<p>By specifying the <code>pattern</code> argument, individual or a
subset of files can be read in. For more information, please refer to
<code><a href="../reference/readImagefromTXT.html">?readImagefromTXT</a></code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="spillover">Spillover correction<a class="anchor" aria-label="anchor" href="#spillover"></a>
</h2>
<p>When acquiring IMC images, pixel intensities can be influenced by
spillover from neighboring channels. To correct for this, Chevrier
<em>et al.</em> have developed a staining protocol to acquire
individually spotted metal isotopes <span class="citation">(Chevrier et
al. 2017)</span>. Based on these measurements, spillover into
neighboring channels can be quantified to correct pixel intensities.</p>
<p>The <code>imcRtools</code> package provides helper functions that
facilitate the correction of spillover for IMC data. For a full
tutorial, please refer to the <a href="https://bodenmillergroup.github.io/IMCDataAnalysis/spillover-correction.html" class="external-link">IMC
data analysis book</a>.</p>
<div class="section level3">
<h3 id="read-in-the-single-spot-acquisitions">Read in the single-spot acquisitions<a class="anchor" aria-label="anchor" href="#read-in-the-single-spot-acquisitions"></a>
</h3>
<p>In the first step, the pixel intensities of individually spotted
metals need to be read into a <code>SingleCellExperiment</code>
container for downstream use with the <em><a href="https://bioconductor.org/packages/3.18/CATALYST" class="external-link">CATALYST</a></em>
package. For this, the <code>readSCEfromTXT</code> function can be
used:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/spillover"</span>, package <span class="op">=</span> <span class="st">"imcRtools"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readSCEfromTXT.html">readSCEfromTXT</a></span><span class="op">(</span><span class="va">path</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Spotted channels:  Dy161, Dy162, Dy163, Dy164</span></span>
<span><span class="co">## Acquired channels:  Dy161, Dy162, Dy163, Dy164</span></span>
<span><span class="co">## Channels spotted but not acquired:  </span></span>
<span><span class="co">## Channels acquired but not spotted:</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 4 400 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(4): 161Dy(Dy161Di) 162Dy(Dy162Di) 163Dy(Dy163Di)</span></span>
<span><span class="co">##   164Dy(Dy164Di)</span></span>
<span><span class="co">## rowData names(2): channel_name marker_name</span></span>
<span><span class="co">## colnames(400): Dy161.1 Dy161.2 ... Dy164.99 Dy164.100</span></span>
<span><span class="co">## colData names(9): Start_push End_push ... sample_metal sample_mass</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>Here, the example metal spot files are read in. The spot information
are stored in the <code>colData(sce)</code> slot and channel information
are stored in <code>rowData(sce)</code>. Each column represents a single
pixel.</p>
</div>
<div class="section level3">
<h3 id="quality-control-on-single-spot-acquisitions">Quality control on single-spot acquisitions<a class="anchor" aria-label="anchor" href="#quality-control-on-single-spot-acquisitions"></a>
</h3>
<p>In the next step, it is crucial to identify potentially mislabeled
spots or spots with low pixel intensities. The <code>imcRtools</code>
package exports the <code>plotSpotHeatmap</code> function, which
visualizes the aggregated (default <code>median</code>) pixel
intensities per spot and per metal:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpotHeatmap.html">plotSpotHeatmap</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/plotSpotHeatmap-1.png" width="480"></p>
<p>Here, high median pixel intensities can be observed in each spot and
their corresponding channels (visualized on the <code>log10</code> scale
by default). To quickly identify spot/channel combinations with low
signal, the <code>threshold</code> parameter can be set:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpotHeatmap.html">plotSpotHeatmap</a></span><span class="op">(</span><span class="va">sce</span>, log <span class="op">=</span> <span class="cn">FALSE</span>, threshold <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/plotSpotHeatmap-2-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="consecutive-pixel-binning">Consecutive pixel binning<a class="anchor" aria-label="anchor" href="#consecutive-pixel-binning"></a>
</h3>
<p>If pixel intensities are low, spillover estimation might not be
robust. Therefore, the <code>binAcrossPixels</code> function can be used
to sum consecutive pixels and enhance the acquired signal. This step is
optional for spillover estimation.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/binAcrossPixels.html">binAcrossPixels</a></span><span class="op">(</span><span class="va">sce</span>, bin_size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSpotHeatmap.html">plotSpotHeatmap</a></span><span class="op">(</span><span class="va">sce2</span>, log <span class="op">=</span> <span class="cn">FALSE</span>, threshold <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/pixel-binning-1.png" width="480"></p>
</div>
<div class="section level3">
<h3 id="pixel-filtering">Pixel filtering<a class="anchor" aria-label="anchor" href="#pixel-filtering"></a>
</h3>
<p>Prior to spillover estimation, the <em><a href="https://bioconductor.org/packages/3.18/CATALYST" class="external-link">CATALYST</a></em>
package provides the <code>assignPrelim</code>, <code>estCutoffs</code>
and <code>applyCutoffs</code> functions to estimate the spotted mass for
each pixel based on their channel intensities. For more information on
the spillover estimation and correction, please refer to the <a href="https://bioconductor.org/packages/release/bioc/vignettes/CATALYST/inst/doc/preprocessing.html#compensation" class="external-link">CATALYST
vignette</a>.</p>
<p>This estimation can be used to identify pixels that cannot be easily
assigned to their spotted mass, potentially indicating pixels with weak
signal. To remove these pixels, the <code>filterPixels</code> function
can be used. This function further removes pixels assigned to masses,
which only contain very few pixels.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HelenaLC/CATALYST" class="external-link">CATALYST</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">bc_key</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">sample_mass</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">assay</span><span class="op">(</span><span class="va">sce</span>, <span class="st">"exprs"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Hyperbolic.html" class="external-link">asinh</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">/</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CATALYST/man/assignPrelim.html" class="external-link">assignPrelim</a></span><span class="op">(</span><span class="va">sce</span>, bc_key <span class="op">=</span> <span class="va">bc_key</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CATALYST/man/estCutoffs.html" class="external-link">estCutoffs</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CATALYST/man/applyCutoffs.html" class="external-link">applyCutoffs</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Filter out mislabeled pixels</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterPixels.html">filterPixels</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">bc_id</span>, <span class="va">sce</span><span class="op">$</span><span class="va">sample_mass</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      </span></span>
<span><span class="co">##       161 162 163 164</span></span>
<span><span class="co">##   0     0   0   0   2</span></span>
<span><span class="co">##   161 100   0   0   0</span></span>
<span><span class="co">##   162   0 100   0   0</span></span>
<span><span class="co">##   163   0   0 100   0</span></span>
<span><span class="co">##   164   0   0   0  98</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="estimating-the-spillover-matrix">Estimating the spillover matrix<a class="anchor" aria-label="anchor" href="#estimating-the-spillover-matrix"></a>
</h3>
<p>Finally, the pre-processed <code>SiingleCellExperiment</code> object
can be used to generate the spillover matrix using the
<code><a href="https://rdrr.io/pkg/CATALYST/man/computeSpillmat.html" class="external-link">CATALYST::computeSpillmat</a></code> function:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CATALYST/man/computeSpillmat.html" class="external-link">computeSpillmat</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu">metadata</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">spillover_matrix</span></span></code></pre></div>
<pre><code><span><span class="co">##             Dy161Di     Dy162Di     Dy163Di     Dy164Di</span></span>
<span><span class="co">## Dy161Di 1.000000000 0.031443129 0.009734712 0.006518048</span></span>
<span><span class="co">## Dy162Di 0.015715159 1.000000000 0.048116187 0.008250039</span></span>
<span><span class="co">## Dy163Di 0.003809504 0.012159704 1.000000000 0.020214651</span></span>
<span><span class="co">## Dy164Di 0.005058069 0.008457546 0.028912343 1.000000000</span></span></code></pre>
<p>This spillover matrix can be directly applied to compensate the
summarized pixel intensities per cell and per channel as described <a href="https://bioconductor.org/packages/release/bioc/vignettes/CATALYST/inst/doc/preprocessing.html#compcytof-compensation-of-mass-cytometry-data" class="external-link">here</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="spatial-analysis">Spatial analysis<a class="anchor" aria-label="anchor" href="#spatial-analysis"></a>
</h2>
<p>The following section will highlight functions for spatial analyses
of the data.</p>
<div class="section level3">
<h3 id="build-graph">Constructing graphs<a class="anchor" aria-label="anchor" href="#build-graph"></a>
</h3>
<p>When following the <code>ImcSegmentationPipeline</code> or
<code>steinbock</code> and reading in the data using the corresponding
functions, the generated graphs are automatically stored in the
<code>colPair(spe, "neighborhood")</code> slot. Alternatively, the
<code>buildSpatialGraph</code> function in the <code>imcRtools</code>
package constructs interaction graphs using either (i) cell-centroid
expansion, (ii) k-nearest neighbor search or (iii) delaunay
triangulation.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/BodenmillerGroup/cytomapper" class="external-link">cytomapper</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pancreasSCE"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildSpatialGraph.html">buildSpatialGraph</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>                                 type <span class="op">=</span> <span class="st">"expansion"</span>,</span>
<span>                                 threshold <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildSpatialGraph.html">buildSpatialGraph</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>                                 type <span class="op">=</span> <span class="st">"knn"</span>,</span>
<span>                                 k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildSpatialGraph.html">buildSpatialGraph</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>                                 type <span class="op">=</span> <span class="st">"delaunay"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colPairs.html" class="external-link">colPairNames</a></span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "expansion_interaction_graph" "knn_interaction_graph"      </span></span>
<span><span class="co">## [3] "delaunay_interaction_graph"</span></span></code></pre>
<p>When setting <code>type = "knn"</code>, by default a directional
graph will be build. Setting <code>directed = FALSE</code> will create
bi-directional edges for each pair of cells that are connected by at
least one edge in the directed setting.</p>
</div>
<div class="section level3">
<h3 id="viz-cells">Graph/cell visualization<a class="anchor" aria-label="anchor" href="#viz-cells"></a>
</h3>
<p>The cells’ locations and constructed graphs can be visualized using
the <code>plotSpatial</code> function. Here, cells are referred to as
“nodes” and cell-cell interactions are referred to as “edges”. All
visual attributes of the nodes and edges can be set. Either by
specifying a variable in <code>colData(spe)</code>, a marker name or a
single entry using the <code>*_fix</code> parameters. By default the
<code>plotSpatial</code> function will visualize equal physical units on
the x- and y-axis with an aspect ratio of 1. The example data are
located in different regions of an image and we therefore set
<code>scales = "free"</code> for simpler visualization.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggraph.data-imaginist.com" class="external-link">ggraph</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>            node_color_by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>            node_shape_by <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>            node_size_by <span class="op">=</span> <span class="st">"Area"</span>,</span>
<span>            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,</span>
<span>            directed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/plotSpatial-1.png" width="700"></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Colored by expression and with arrows</span></span>
<span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>            node_color_by <span class="op">=</span> <span class="st">"PIN"</span>,</span>
<span>            assay_type <span class="op">=</span> <span class="st">"exprs"</span>,</span>
<span>            node_size_fix <span class="op">=</span> <span class="fl">3</span>,</span>
<span>            edge_width_fix <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,</span>
<span>            directed <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            arrow <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/arrow.html" class="external-link">arrow</a></span><span class="op">(</span>length <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="st">"inch"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            end_cap <span class="op">=</span> <span class="fu">ggraph</span><span class="fu">::</span><span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geometry.html" class="external-link">circle</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>            scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/plotSpatial-2.png" width="700"></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Subsetting the SingleCellExperiment</span></span>
<span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">[</span>,<span class="va">pancreasSCE</span><span class="op">$</span><span class="va">Pattern</span><span class="op">]</span>,</span>
<span>            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>            node_color_by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>            node_size_fix <span class="op">=</span> <span class="fl">1</span>,</span>
<span>            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,</span>
<span>            directed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span></code></pre></div>
<p><img src="imcRtools_files/figure-html/plotSpatial-3.png" width="700"></p>
<p>The returned object can be further modified using the
<code>ggplot2</code> logic. This includes changing the node color, shape
and size using <code>scale_color_*</code>, <code>scale_shape_*</code>
and <code>scale_size_*</code>. Edge attributes can be altered using the
<code>scale_edge_*</code> function exported by <code>ggraph</code>,</p>
</div>
<div class="section level3">
<h3 id="aggregate-neighbors">Neighborhood aggregation<a class="anchor" aria-label="anchor" href="#aggregate-neighbors"></a>
</h3>
<p>The <code>aggregateNeighbors</code> function can be used to aggregate
features of all neighboring cells for each individual cell. This
function operates in two settings.</p>
<ol style="list-style-type: decimal">
<li>
<code>metadata</code>: when aggregating by cell-specific metadata,
the function computes the relative frequencies of all entries to
<code>colData(sce)[[count_by]]</code> within the direct neighborhood of
each cell.<br>
</li>
<li>
<code>expression</code>: the expression counts of neighboring cells
are aggregated using the specified <code>statistic</code> (defaults to
<code>mean</code>).</li>
</ol>
<p>Each cell’s neighborhood is defined as endpoints of edges stored in
<code>colPair(sce, colPairName)</code>.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregateNeighbors.html">aggregateNeighbors</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>                                  colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,</span>
<span>                                  aggregate_by <span class="op">=</span> <span class="st">"metadata"</span>,</span>
<span>                                  count_by <span class="op">=</span> <span class="st">"CellType"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">$</span><span class="va">aggregatedNeighbors</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 3 columns</span></span>
<span><span class="co">##   celltype_A celltype_B celltype_C</span></span>
<span><span class="co">##    &lt;numeric&gt;  &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">## 1          0        0.0        1.0</span></span>
<span><span class="co">## 2          0        0.2        0.8</span></span>
<span><span class="co">## 3          0        0.0        1.0</span></span>
<span><span class="co">## 4          0        0.0        1.0</span></span>
<span><span class="co">## 5          0        0.0        1.0</span></span>
<span><span class="co">## 6          0        0.0        1.0</span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregateNeighbors.html">aggregateNeighbors</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>                                  colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,</span>
<span>                                  aggregate_by <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>                                  assay_type <span class="op">=</span>  <span class="st">"exprs"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">$</span><span class="va">mean_aggregatedExpression</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 5 columns</span></span>
<span><span class="co">##          H3      CD99       PIN      CD8a       CDH</span></span>
<span><span class="co">##   &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1   2.32500  0.860329  0.092871  0.725000   2.51264</span></span>
<span><span class="co">## 2   2.88022  1.629762  0.319527  0.207873   2.46486</span></span>
<span><span class="co">## 3   3.10829  0.735389  0.190616  0.255515   1.89484</span></span>
<span><span class="co">## 4   2.55842  0.773342  0.124545  0.188629   2.51084</span></span>
<span><span class="co">## 5   2.44287  1.126240  0.252129  0.200261   2.61336</span></span>
<span><span class="co">## 6   2.65059  0.903869  0.181792  0.196691   2.16434</span></span></code></pre>
<p>The returned entries can now be used for clustering to group cells
based on their environment (either by aggregated categorical features or
expression).</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">22</span><span class="op">)</span></span>
<span><span class="va">cur_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">$</span><span class="va">aggregatedNeighbors</span>, centers <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">pancreasSCE</span><span class="op">$</span><span class="va">clustered_neighbors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cur_cluster</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize CellType and clustered_neighbors</span></span>
<span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>            node_color_by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>            node_size_fix <span class="op">=</span> <span class="fl">4</span>,</span>
<span>            edge_width_fix <span class="op">=</span> <span class="fl">2</span>,</span>
<span>            edge_color_by <span class="op">=</span> <span class="st">"clustered_neighbors"</span>,</span>
<span>            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,</span>
<span>            directed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            nodes_first <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_colour.html" class="external-link">scale_edge_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/aggregateNeigbors-clustering-1.png" width="700"></p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualize clustered_neighbors</span></span>
<span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>            node_color_by <span class="op">=</span> <span class="st">"clustered_neighbors"</span>,</span>
<span>            node_size_fix <span class="op">=</span> <span class="fl">4</span>,</span>
<span>            edge_width_fix <span class="op">=</span> <span class="fl">1</span>,</span>
<span>            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,</span>
<span>            directed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            nodes_first <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span><span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/aggregateNeigbors-clustering-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="spatial-context">Spatial context analysis<a class="anchor" aria-label="anchor" href="#spatial-context"></a>
</h3>
<p>The single cell assignments derived from clustering cells based on
their environment can be interpreted as <strong>cellular neighborhoods
(CNs)</strong>, which can represent sites of unique local processes
<span class="citation">(Schürch et al. 2020)</span>.</p>
<p>Downstream of CNs, imcRtools exports three functions to detect and
analyze the <strong>spatial context (SC)</strong> of each cell.</p>
<ol style="list-style-type: decimal">
<li>
<code>detectSpatialContext</code>: for the function to detect
SCs</li>
<li>
<code>filterSpatialContext</code>: for the function to filter
SCs</li>
<li>
<code>plotSpatialContext</code>: for the function to plot SC
graphs</li>
</ol>
<p>The term SC was coined by Bhate and colleagues <span class="citation">(Bhate et al. 2022)</span> and describes tissue regions
in which distinct CNs may be interacting, which can lead to specialized
local biological events.</p>
<p>The <code>detectSpatialContext</code> function relies on CN fractions
for each cell in a spatial interaction graph (originally a k-nearest
neighbor (KNN) graph).</p>
<p>We can retrieve the CN fractions using the above-described
<code>buildSpatialGraph</code> and <code>aggregateNeighbors</code>
functions. The window size (k for KNN) for
<code>buildSpatialGraph</code> should reflect a length scale on which
biological signals can be exchanged and depends, among others, on cell
density and tissue area. In view of their divergent functionality, we
recommend to use a larger window size for SC (interaction between local
processes) than for CN (local processes) detection.</p>
<p>Subsequently, the CN fractions are sorted from high-to-low and the SC
of each cell is assigned the minimal combination of SCs that additively
surpass a user-defined threshold. The default threshold of 0.9 aims to
represent the dominant CNs, hence the most prevalent signals, in a given
window.</p>
<p>For more details, please refer to <span class="citation">(Bhate et
al. 2022)</span>.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate k-nearest neighbor graph</span></span>
<span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildSpatialGraph.html">buildSpatialGraph</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>                                 type <span class="op">=</span> <span class="st">"knn"</span>,</span>
<span>                                 name <span class="op">=</span> <span class="st">"knn_spatialcontext_graph"</span>,</span>
<span>                                 k <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colPairs.html" class="external-link">colPairNames</a></span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "expansion_interaction_graph" "knn_interaction_graph"      </span></span>
<span><span class="co">## [3] "delaunay_interaction_graph"  "knn_spatialcontext_graph"</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Aggregate based on clustered_neighbors</span></span>
<span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggregateNeighbors.html">aggregateNeighbors</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, </span>
<span>                                  colPairName <span class="op">=</span> <span class="st">"knn_spatialcontext_graph"</span>,</span>
<span>                                  aggregate_by <span class="op">=</span> <span class="st">"metadata"</span>,</span>
<span>                                  count_by <span class="op">=</span> <span class="st">"clustered_neighbors"</span>,</span>
<span>                                  name <span class="op">=</span> <span class="st">"aggregatedNeighborhood"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Detect spatial contexts</span></span>
<span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectSpatialContext.html">detectSpatialContext</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, </span>
<span>                                    entry <span class="op">=</span> <span class="st">"aggregatedNeighborhood"</span>,</span>
<span>                                    threshold <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>                                    name <span class="op">=</span> <span class="st">"spatial_context"</span><span class="op">)</span></span>
<span><span class="co"># Define SC color scheme</span></span>
<span><span class="va">col_SC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#A6CEE3"</span>, <span class="st">"#1F78B4"</span>, <span class="st">"#B2DF8A"</span>, <span class="st">"#33A02C"</span>, <span class="st">"#FB9A99"</span>, <span class="st">"#E31A1C"</span>, <span class="st">"#FDBF6F"</span><span class="op">)</span>, </span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">$</span><span class="va">spatial_context</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Visualize spatial contexts on images</span></span>
<span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>            node_color_by <span class="op">=</span> <span class="st">"spatial_context"</span>,</span>
<span>            node_size_fix <span class="op">=</span> <span class="fl">4</span>,</span>
<span>            edge_width_fix <span class="op">=</span> <span class="fl">1</span>,</span>
<span>            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colPairName <span class="op">=</span> <span class="st">"knn_spatialcontext_graph"</span>,</span>
<span>            directed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            nodes_first <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">col_SC</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/detectSpatialContext-1.png" width="700"></p>
<p>After SC assignment for each individual cell, the
<code>filterSpatialContext</code> function allows to filter detected SCs
based on user-defined thresholds for number of group entries (usually
image or patient ID) and/or total number of cells per SC.</p>
<p>In addition to a new column entry to the
<code>colData(object)</code>, the function also returns a
<code>data.frame</code> entry to <code>metadata(object)</code>
containing filtered group and cell counts per SC.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Filter spatial contexts</span></span>
<span><span class="co"># By number of group entries</span></span>
<span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterSpatialContext.html">filterSpatialContext</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, </span>
<span>                            entry <span class="op">=</span> <span class="st">"spatial_context"</span>,</span>
<span>                            group_by <span class="op">=</span> <span class="st">"ImageNb"</span>, </span>
<span>                            group_threshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                            name <span class="op">=</span> <span class="st">"spatial_context_filtered_group"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">metadata</span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">)</span><span class="op">$</span><span class="va">filterSpatialContext</span></span></code></pre></div>
<pre><code><span><span class="co">##   spatial_context n_cells n_group</span></span>
<span><span class="co">## 1               1      87       2</span></span>
<span><span class="co">## 2             1_2      71       2</span></span>
<span><span class="co">## 4             1_3      90       2</span></span>
<span><span class="co">## 5               2      55       2</span></span>
<span><span class="co">## 7               3      14       2</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>            node_color_by <span class="op">=</span> <span class="st">"spatial_context_filtered_group"</span>,</span>
<span>            node_size_fix <span class="op">=</span> <span class="fl">4</span>,</span>
<span>            edge_width_fix <span class="op">=</span> <span class="fl">1</span>,</span>
<span>            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colPairName <span class="op">=</span> <span class="st">"knn_spatialcontext_graph"</span>,</span>
<span>            directed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            nodes_first <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">col_SC</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/filterSpatialContext-1.png" width="700"></p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># By total number of cells</span></span>
<span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterSpatialContext.html">filterSpatialContext</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, </span>
<span>                            entry <span class="op">=</span> <span class="st">"spatial_context"</span>,</span>
<span>                            group_by <span class="op">=</span> <span class="st">"ImageNb"</span>, </span>
<span>                            cells_threshold <span class="op">=</span> <span class="fl">15</span>,</span>
<span>                            name <span class="op">=</span> <span class="st">"spatial_context_filtered_cells"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">metadata</span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">)</span><span class="op">$</span><span class="va">filterSpatialContext</span></span></code></pre></div>
<pre><code><span><span class="co">##   spatial_context n_cells n_group</span></span>
<span><span class="co">## 1               1      87       2</span></span>
<span><span class="co">## 2             1_2      71       2</span></span>
<span><span class="co">## 3           1_2_3      16       1</span></span>
<span><span class="co">## 4             1_3      90       2</span></span>
<span><span class="co">## 5               2      55       2</span></span>
<span><span class="co">## 6             2_3      29       1</span></span></code></pre>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>            node_color_by <span class="op">=</span> <span class="st">"spatial_context_filtered_cells"</span>,</span>
<span>            node_size_fix <span class="op">=</span> <span class="fl">4</span>,</span>
<span>            edge_width_fix <span class="op">=</span> <span class="fl">1</span>,</span>
<span>            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colPairName <span class="op">=</span> <span class="st">"knn_spatialcontext_graph"</span>,</span>
<span>            directed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            nodes_first <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">col_SC</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/filterSpatialContext-2.png" width="700"></p>
<p>Lastly, the <code>plotSpatialContext</code> function plots directed
<em>SC graphs</em>, akin to <em>CN combination maps</em> in <span class="citation">(Bhate et al. 2022)</span>, based on symbolic
edge-lists and vertex metadata, which operates on cohort-level.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot spatial context graph </span></span>
<span><span class="co"># Default</span></span>
<span><span class="fu"><a href="../reference/plotSpatialContext.html">plotSpatialContext</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>                   entry <span class="op">=</span> <span class="st">"spatial_context"</span>, </span>
<span>                   group_by <span class="op">=</span> <span class="st">"ImageNb"</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/plotSpatialContext-1.png" width="700"></p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Colored by name and size by n_cells</span></span>
<span><span class="fu"><a href="../reference/plotSpatialContext.html">plotSpatialContext</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, </span>
<span>                   entry <span class="op">=</span> <span class="st">"spatial_context"</span>,</span>
<span>                   group_by <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>                   node_color_by <span class="op">=</span> <span class="st">"name"</span>,</span>
<span>                   node_size_by <span class="op">=</span> <span class="st">"n_cells"</span>,</span>
<span>                   node_label_color_by <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/plotSpatialContext-2.png" width="700"></p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Colored by n_cells and size by n_group                   </span></span>
<span><span class="fu"><a href="../reference/plotSpatialContext.html">plotSpatialContext</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, </span>
<span>                   entry <span class="op">=</span> <span class="st">"spatial_context"</span>,</span>
<span>                   group_by <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>                   node_color_by <span class="op">=</span> <span class="st">"n_cells"</span>,</span>
<span>                   node_size_by <span class="op">=</span> <span class="st">"n_group"</span>,</span>
<span>                   node_label_color_by <span class="op">=</span> <span class="st">"n_cells"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_color_viridis</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/plotSpatialContext-3.png" width="700"></p>
<p>The returned object can be further modified using the
<code>ggplot2</code> logic. This includes changing the node color and
size using <code>scale_color_*</code> and <code>scale_size_*</code>.
Edge attributes can be altered using the <code>scale_edge_*</code>
function exported by <code>ggraph</code>.</p>
<p>Furthermore, setting <code>return_data = TRUE</code> returns the
symbolic edge-list and vertex metadata used for graph construction in a
<code>list</code> of two <code>data.frames</code>.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Return data</span></span>
<span><span class="fu"><a href="../reference/plotSpatialContext.html">plotSpatialContext</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>                   entry <span class="op">=</span> <span class="st">"spatial_context"</span>, </span>
<span>                   group_by <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>                   return_data <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $edges</span></span>
<span><span class="co">##   from    to</span></span>
<span><span class="co">## 1    1   1_2</span></span>
<span><span class="co">## 2    1   1_3</span></span>
<span><span class="co">## 3  1_2 1_2_3</span></span>
<span><span class="co">## 4  1_3 1_2_3</span></span>
<span><span class="co">## 5    2   1_2</span></span>
<span><span class="co">## 6    2   2_3</span></span>
<span><span class="co">## 7  2_3 1_2_3</span></span>
<span><span class="co">## 8    3   1_3</span></span>
<span><span class="co">## 9    3   2_3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $vertices</span></span>
<span><span class="co">##   spatial_context n_cells n_group length</span></span>
<span><span class="co">## 1               1      87       2      1</span></span>
<span><span class="co">## 2             1_2      71       2      2</span></span>
<span><span class="co">## 3           1_2_3      16       1      3</span></span>
<span><span class="co">## 4             1_3      90       2      2</span></span>
<span><span class="co">## 5               2      55       2      1</span></span>
<span><span class="co">## 6             2_3      29       1      2</span></span>
<span><span class="co">## 7               3      14       2      1</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="community-detection">Community detection<a class="anchor" aria-label="anchor" href="#community-detection"></a>
</h3>
<p>In addition to cellular neighborhood and spatial context analysis,
<code>imcRtools</code> exports the <code>detectCommunity</code> function
to detect the spatial community of each cell as proposed by <span class="citation">(Jackson et al. 2020)</span>. Here, each cell is
clustered based on its interactions as defined by a spatial object
graph.</p>
<p>In more detail, the spatial community detection procedure is as
follows:</p>
<ol style="list-style-type: decimal">
<li><p>Create an igraph object from the edge list stored in .</p></li>
<li><p>Perform community detection using the specified algorithm
(defaults to “louvain”).</p></li>
<li><p>Store the community IDs in a vector and replace all communities
with a size smaller than by NA.</p></li>
</ol>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Detect spatial community </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">22</span><span class="op">)</span></span>
<span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectCommunity.html">detectCommunity</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, </span>
<span>                               colPairName <span class="op">=</span> <span class="st">"expansion_interaction_graph"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>            node_color_by <span class="op">=</span> <span class="st">"spatial_community"</span>,</span>
<span>            node_size_fix <span class="op">=</span> <span class="fl">4</span>,</span>
<span>            edge_width_fix <span class="op">=</span> <span class="fl">1</span>,</span>
<span>            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colPairName <span class="op">=</span> <span class="st">"expansion_interaction_graph"</span>,</span>
<span>            directed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            nodes_first <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/detectCommunity-1.png" width="700"></p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Detect spatial community - specify size_threshold</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">22</span><span class="op">)</span></span>
<span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectCommunity.html">detectCommunity</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, </span>
<span>                               colPairName <span class="op">=</span> <span class="st">"expansion_interaction_graph"</span>, </span>
<span>                               size_threshold <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>            node_color_by <span class="op">=</span> <span class="st">"spatial_community"</span>,</span>
<span>            node_size_fix <span class="op">=</span> <span class="fl">4</span>,</span>
<span>            edge_width_fix <span class="op">=</span> <span class="fl">1</span>,</span>
<span>            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colPairName <span class="op">=</span> <span class="st">"expansion_interaction_graph"</span>,</span>
<span>            directed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            nodes_first <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/detectCommunity-2.png" width="700">
It is also possible to use different community detection algorithms from
the <code>igraph</code>package.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Detect spatial community - walktrap community detection</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">22</span><span class="op">)</span></span>
<span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectCommunity.html">detectCommunity</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, </span>
<span>                               colPairName <span class="op">=</span> <span class="st">"expansion_interaction_graph"</span>, </span>
<span>                               cluster_fun <span class="op">=</span> <span class="st">"walktrap"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>            node_color_by <span class="op">=</span> <span class="st">"spatial_community"</span>,</span>
<span>            node_size_fix <span class="op">=</span> <span class="fl">4</span>,</span>
<span>            edge_width_fix <span class="op">=</span> <span class="fl">1</span>,</span>
<span>            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colPairName <span class="op">=</span> <span class="st">"expansion_interaction_graph"</span>,</span>
<span>            directed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            nodes_first <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/detectCommunity-walktrap-1.png" width="700"></p>
<p>Moreover, the user can specify to perform spatial community detection
separately for all unique entries to e.g. for tumor and non-tumor
cells.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Detect spatial community - specify group_by</span></span>
<span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detectCommunity.html">detectCommunity</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, </span>
<span>                               colPairName <span class="op">=</span> <span class="st">"expansion_interaction_graph"</span>, </span>
<span>                               group_by <span class="op">=</span> <span class="st">"CellType"</span>, </span>
<span>                               size_threshold <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                               BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">SerialParam</a></span><span class="op">(</span>RNGseed <span class="op">=</span> <span class="fl">22</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>            node_color_by <span class="op">=</span> <span class="st">"spatial_community"</span>,</span>
<span>            node_size_fix <span class="op">=</span> <span class="fl">4</span>,</span>
<span>            edge_width_fix <span class="op">=</span> <span class="fl">1</span>,</span>
<span>            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colPairName <span class="op">=</span> <span class="st">"expansion_interaction_graph"</span>,</span>
<span>            directed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            nodes_first <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/detectCommunity-bygroup-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="border-cells">Border cells<a class="anchor" aria-label="anchor" href="#border-cells"></a>
</h3>
<p>To exclude cells that are close to the image border, the
<code>imcRtools</code> package exports the <code>findBorderCells</code>
function.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findBorderCells.html">findBorderCells</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>                               img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>                               border_dist <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">[</span>,<span class="op">!</span><span class="va">pancreasSCE</span><span class="op">$</span><span class="va">border_cells</span><span class="op">]</span>,</span>
<span>            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>            node_color_by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>            node_size_fix <span class="op">=</span> <span class="fl">4</span>,</span>
<span>            edge_width_fix <span class="op">=</span> <span class="fl">2</span>,</span>
<span>            edge_color_by <span class="op">=</span> <span class="st">"clustered_neighbors"</span>,</span>
<span>            draw_edges <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,</span>
<span>            directed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            nodes_first <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_colour.html" class="external-link">scale_edge_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/findBorderCells-1.png" width="700"></p>
<p>Excluding border cells can be useful when incorrectly connected cells
are observed at image borders.</p>
</div>
<div class="section level3">
<h3 id="patch-detection">Patch detection<a class="anchor" aria-label="anchor" href="#patch-detection"></a>
</h3>
<p>An alternative and supervised way of detecting regions with similar
types of cells is available via the <code>patchDetection</code>
function. Here, the user defines which cells should be used for patch
detection via the <code>patch_cells</code> parameter. A patch is defined
as a set of cells as defined by <code>patch_cells</code> which are
weakly connected in the graph in
<code>colPair(object, colPairname)</code>.</p>
<p>Below, the <code>patchDetection</code> function is demonstrated using
the previously computed <code>expansion</code> graph and defining cells
of <code>celltype_B</code> as the cells of interest. Here, the function
additionally draws a concave hull around the detected patch, expands the
hull by 20<span class="math inline">\(\mu{}m\)</span> and defines all
cells within this expanded hulls as patch cells.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/patchDetection.html">patchDetection</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, </span>
<span>                              patch_cells <span class="op">=</span> <span class="va">pancreasSCE</span><span class="op">$</span><span class="va">CellType</span> <span class="op">==</span> <span class="st">"celltype_B"</span>,</span>
<span>                              colPairName <span class="op">=</span> <span class="st">"expansion_interaction_graph"</span>,</span>
<span>                              expand_by <span class="op">=</span> <span class="fl">20</span>, </span>
<span>                              img_id <span class="op">=</span> <span class="st">"ImageNb"</span><span class="op">)</span></span>
<span> </span>
<span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>, </span>
<span>            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>, </span>
<span>            node_color_by <span class="op">=</span> <span class="st">"patch_id"</span>,</span>
<span>            scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/patchDetection-1.png" width="700"></p>
<p>Patches that only consist of 1 or 2 cells cannot be expanded.</p>
<p>For each patch larger than 2 cells, the spatial area can be computed
using the <code>patchSize</code> function:</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/patchSize.html">patchSize</a></span><span class="op">(</span><span class="va">pancreasSCE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 8 rows and 3 columns</span></span>
<span><span class="co">##      patch_id                polygon      size</span></span>
<span><span class="co">##   &lt;character&gt;                 &lt;list&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1           1                     NA        NA</span></span>
<span><span class="co">## 2           2                     NA        NA</span></span>
<span><span class="co">## 3           3 c(180.4247, 170.4, 1..   913.849</span></span>
<span><span class="co">## 4           4                     NA        NA</span></span>
<span><span class="co">## 5           5                     NA        NA</span></span>
<span><span class="co">## 6           6 c(138.08, 134.4638, ..  4066.407</span></span>
<span><span class="co">## 7           7 c(133.1014, 133.9213..  1283.972</span></span>
<span><span class="co">## 8           8 c(244.3196, 247.878,..  6224.065</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="minimal-distances-to-cells-of-interest">Minimal distances to cells of interest<a class="anchor" aria-label="anchor" href="#minimal-distances-to-cells-of-interest"></a>
</h3>
<p>Calculate the minimal distance for each cell to a given cell type or
class of cells of interest, the function <code>minDistToCells</code> is
available. Cells of interest are defined via the <code>x_cells</code>
parameter as <code>logical</code> and distances to for all cells to
those cells will be reported in a new column in the <code>colData</code>
of the <code>SingleCellExperiment</code>.</p>
<p>If the cells of interest form patches (many cells of the same type
next to each other) or similarly if a patch detection has previously
been performed the positive distances reflect the distances from cells
outside of patches to the closest patch border and the negative
distances reflect the distances from cells inside the patches to the
patch border. If <code>return_neg</code> is set to <code>FALSE</code>
negative distances are set to 0.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pancreasSCE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/minDistToCells.html">minDistToCells</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>                              x_cells <span class="op">=</span> <span class="va">pancreasSCE</span><span class="op">$</span><span class="va">CellType</span> <span class="op">==</span> <span class="st">"celltype_B"</span>,</span>
<span>                              coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Pos_X"</span>,<span class="st">"Pos_Y"</span><span class="op">)</span>,</span>
<span>                              img_id <span class="op">=</span> <span class="st">"ImageNb"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSpatial.html">plotSpatial</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>            img_id <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>            node_color_by <span class="op">=</span> <span class="st">"distToCells"</span>,</span>
<span>            scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_color_viridis</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="imcRtools_files/figure-html/minDistToCells-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="test-neighborhood">Neighborhood permutation testing<a class="anchor" aria-label="anchor" href="#test-neighborhood"></a>
</h3>
<p>The following section describes how to observe and test the average
number of interactions between cell labels (e.g. cell-types) within
grouping levels (e.g. images). For full descriptions of the testing
approaches, please refer to <a href="https://www.nature.com/articles/nmeth.4391" class="external-link">Shapiro et al., Nature
Methods</a> <span class="citation">(Schapiro et al. 2017)</span> and <a href="https://www.sciencedirect.com/science/article/pii/S2405471217305434" class="external-link">Schulz
et al., Cell Systems</a> <span class="citation">(Schulz et al.
2018)</span></p>
<p>The <code>imcRtools</code> package exports the
<code>countInteractions</code> and <code>testInteractions</code>
functions, which summarize all cell-cell interactions per grouping level
(e.g. image). As a result, a table is returned where each row represents
one of all possible cell-type/cell-type interactions among all grouping
levels. Missing entries or <code>NA</code>s indicate missing cell-type
labels for this grouping level. The next section gives details on how
interactions are summarized.</p>
<div class="section level4">
<h4 id="summarizing-interactions">Summarizing interactions<a class="anchor" aria-label="anchor" href="#summarizing-interactions"></a>
</h4>
<p>The <code>countInteractions</code> function counts the number of
edges (interactions) between each set of unique cell labels per grouping
level. Simplified, it counts for each cell of type A the number of
neighbors of type B. This count is averaged within each unique grouping
level (e.g. image) in three different ways:</p>
<ol style="list-style-type: decimal">
<li><p><code>method = "classic"</code>: The count is divided by the
total number of cells of type A. The final count can be interpreted as
“How many neighbors of type B does a cell of type A have on
average?”</p></li>
<li><p><code>method = "histocat"</code>: The count is divided by the
number of cells of type A that have at least one neighbor of type B. The
final count can be interpreted as “How many many neighbors of type B has
a cell of type A on average, given it has at least one neighbor of type
B?”</p></li>
<li><p><code>method = "patch"</code>: For each cell, the count is
binarized to 0 (less than <code>patch_size</code> neighbors of type B)
or 1 (more or equal to <code>patch_size</code> neighbors of type B). The
binarized counts are averaged across all cells of type A. The final
count can be interpreted as “What fraction of cells of type A have at
least a given number of neighbors of type B?”</p></li>
</ol>
<p>The <code>countInteractions</code> returns a <code>DataFrame</code>
containing the summarized counts (<code>ct</code>) for all combinations
of <code>from_label</code>, <code>to_label</code> and
<code>group_by</code>.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/countInteractions.html">countInteractions</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>                         group_by <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>                         label <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>                         method <span class="op">=</span> <span class="st">"classic"</span>,</span>
<span>                         colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span><span class="op">)</span></span>
<span><span class="va">out</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 27 rows and 4 columns</span></span>
<span><span class="co">##      group_by from_label   to_label        ct</span></span>
<span><span class="co">##     &lt;integer&gt;   &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1           1 celltype_A celltype_A  2.823529</span></span>
<span><span class="co">## 2           1 celltype_A celltype_B  0.823529</span></span>
<span><span class="co">## 3           1 celltype_A celltype_C  1.352941</span></span>
<span><span class="co">## 4           1 celltype_B celltype_A  2.000000</span></span>
<span><span class="co">## 5           1 celltype_B celltype_B  0.625000</span></span>
<span><span class="co">## ...       ...        ...        ...       ...</span></span>
<span><span class="co">## 23          3 celltype_B celltype_B   4.00000</span></span>
<span><span class="co">## 24          3 celltype_B celltype_C   1.00000</span></span>
<span><span class="co">## 25          3 celltype_C celltype_A        NA</span></span>
<span><span class="co">## 26          3 celltype_C celltype_B   1.13115</span></span>
<span><span class="co">## 27          3 celltype_C celltype_C   3.86885</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="testing-for-significance">Testing for significance<a class="anchor" aria-label="anchor" href="#testing-for-significance"></a>
</h4>
<p>In the next instance, one can test if the obtained count is larger or
smaller compared to what is expected from a random distribution of cell
labels. For this, the <code>testInteractions</code> function permutes
the cell labels <code>iter</code> times and counts interactions as
described above. This approach generates a distribution of the
interaction count under a random distribution of cell labels. The
observed interaction count is compared against this Null distribution to
derive empirical p-values:</p>
<p><code>p_gt</code>: fraction of perturbations equal or greater than
the observed count</p>
<p><code>p_lt</code>: fraction of perturbations equal or less than the
observed count</p>
<p>Based on these empirical p-values, the <code>interaction</code> score
(attraction or avoidance), overall <code>p</code> value and significance
by comparison to <code>p_treshold</code> (<code>sig</code> and
<code>sigval</code>) are derived. All results are returned in form of a
<code>DataFrame</code>.</p>
<p>We set the seed within the <code>SerialParam</code> (or
<code>MulticoreParam</code>) function for reproducibility.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/testInteractions.html">testInteractions</a></span><span class="op">(</span><span class="va">pancreasSCE</span>,</span>
<span>                        group_by <span class="op">=</span> <span class="st">"ImageNb"</span>,</span>
<span>                        label <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>                        method <span class="op">=</span> <span class="st">"classic"</span>,</span>
<span>                        colPairName <span class="op">=</span> <span class="st">"knn_interaction_graph"</span>,</span>
<span>                        BPPARAM <span class="op">=</span> <span class="fu">BiocParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SerialParam-class.html" class="external-link">SerialParam</a></span><span class="op">(</span>RNGseed <span class="op">=</span> <span class="fl">123</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">out</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 27 rows and 10 columns</span></span>
<span><span class="co">##        group_by  from_label    to_label        ct        p_gt        p_lt</span></span>
<span><span class="co">##     &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">## 1             1  celltype_A  celltype_A  2.823529 0.000999001 1.000000000</span></span>
<span><span class="co">## 2             1  celltype_A  celltype_B  0.823529 0.001998002 1.000000000</span></span>
<span><span class="co">## 3             1  celltype_A  celltype_C  1.352941 1.000000000 0.000999001</span></span>
<span><span class="co">## 4             1  celltype_B  celltype_A  2.000000 0.000999001 1.000000000</span></span>
<span><span class="co">## 5             1  celltype_B  celltype_B  0.625000 0.154845155 0.911088911</span></span>
<span><span class="co">## ...         ...         ...         ...       ...         ...         ...</span></span>
<span><span class="co">## 23            3  celltype_B  celltype_B   4.00000 0.000999001 1.000000000</span></span>
<span><span class="co">## 24            3  celltype_B  celltype_C   1.00000 1.000000000 0.000999001</span></span>
<span><span class="co">## 25            3  celltype_C  celltype_A        NA          NA          NA</span></span>
<span><span class="co">## 26            3  celltype_C  celltype_B   1.13115 1.000000000 0.000999001</span></span>
<span><span class="co">## 27            3  celltype_C  celltype_C   3.86885 0.000999001 1.000000000</span></span>
<span><span class="co">##     interaction           p       sig    sigval</span></span>
<span><span class="co">##       &lt;logical&gt;   &lt;numeric&gt; &lt;logical&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1          TRUE 0.000999001      TRUE         1</span></span>
<span><span class="co">## 2          TRUE 0.001998002      TRUE         1</span></span>
<span><span class="co">## 3         FALSE 0.000999001      TRUE        -1</span></span>
<span><span class="co">## 4          TRUE 0.000999001      TRUE         1</span></span>
<span><span class="co">## 5          TRUE 0.154845155     FALSE         0</span></span>
<span><span class="co">## ...         ...         ...       ...       ...</span></span>
<span><span class="co">## 23         TRUE 0.000999001      TRUE         1</span></span>
<span><span class="co">## 24        FALSE 0.000999001      TRUE        -1</span></span>
<span><span class="co">## 25           NA          NA        NA        NA</span></span>
<span><span class="co">## 26        FALSE 0.000999001      TRUE        -1</span></span>
<span><span class="co">## 27         TRUE 0.000999001      TRUE         1</span></span></code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="contributions">Contributions<a class="anchor" aria-label="anchor" href="#contributions"></a>
</h2>
<p>Large chunks of the code of the <code>imcRtools</code> package is
based on the work of <a href="https://github.com/votti" class="external-link">Vito
Zanotelli</a>. Vito has co-developed the spillover correction approach
and translated the interaction testing approach developed by <a href="https://github.com/DenisSch" class="external-link">Denis Shapiro</a> and <a href="https://github.com/JanaFischer" class="external-link">Jana Fischer</a> into R (formerly
known as the <a href="https://github.com/BodenmillerGroup/neighbouRhood" class="external-link">neighbouRhood</a>
R package). Jana has furthermore added the “patch” method for
interaction counting and testing. <a href="https://github.com/toobiwankenobi" class="external-link">Tobias Hoch</a> has written the
first version of reading in the <code>ImcSegmentationPipeline</code>
output and the <code>patchDetection</code> function. <a href="https://github.com/SchulzDan" class="external-link">Daniel Schulz</a> has build the
<code>aggregateNeighbors</code> and <code>minDistToCells</code>
functions and contributed to developing the spatial clustering approach.
<a href="https://github.com/lassedochreden" class="external-link">Lasse Meyer</a> has
implemented the functions for spatial context analysis.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">## R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Monterey 12.6.9</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] ggraph_2.1.0                ggplot2_3.4.4              </span></span>
<span><span class="co">##  [3] cytomapper_1.14.0           EBImage_4.44.0             </span></span>
<span><span class="co">##  [5] CATALYST_1.26.0             imcRtools_1.9.0            </span></span>
<span><span class="co">##  [7] SpatialExperiment_1.12.0    SingleCellExperiment_1.24.0</span></span>
<span><span class="co">##  [9] SummarizedExperiment_1.32.0 Biobase_2.62.0             </span></span>
<span><span class="co">## [11] GenomicRanges_1.54.1        GenomeInfoDb_1.38.1        </span></span>
<span><span class="co">## [13] IRanges_2.36.0              S4Vectors_0.40.1           </span></span>
<span><span class="co">## [15] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      </span></span>
<span><span class="co">## [17] matrixStats_1.1.0           BiocStyle_2.30.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.3                    bitops_1.0-7               </span></span>
<span><span class="co">##   [3] sf_1.0-14                   RColorBrewer_1.1-3         </span></span>
<span><span class="co">##   [5] doParallel_1.0.17           tools_4.3.2                </span></span>
<span><span class="co">##   [7] backports_1.4.1             utf8_1.2.4                 </span></span>
<span><span class="co">##   [9] R6_2.5.1                    DT_0.30                    </span></span>
<span><span class="co">##  [11] HDF5Array_1.30.0            rhdf5filters_1.14.1        </span></span>
<span><span class="co">##  [13] GetoptLong_1.0.5            withr_2.5.2                </span></span>
<span><span class="co">##  [15] sp_2.1-1                    gridExtra_2.3              </span></span>
<span><span class="co">##  [17] cli_3.6.1                   textshaping_0.3.7          </span></span>
<span><span class="co">##  [19] sandwich_3.0-2              labeling_0.4.3             </span></span>
<span><span class="co">##  [21] sass_0.4.7                  nnls_1.5                   </span></span>
<span><span class="co">##  [23] mvtnorm_1.2-3               readr_2.1.4                </span></span>
<span><span class="co">##  [25] ggridges_0.5.4              proxy_0.4-27               </span></span>
<span><span class="co">##  [27] pkgdown_2.0.7               systemfonts_1.0.5          </span></span>
<span><span class="co">##  [29] colorRamps_2.3.1            svglite_2.1.2              </span></span>
<span><span class="co">##  [31] scater_1.30.1               plotrix_3.8-4              </span></span>
<span><span class="co">##  [33] flowCore_2.14.0             generics_0.1.3             </span></span>
<span><span class="co">##  [35] shape_1.4.6                 gtools_3.9.4               </span></span>
<span><span class="co">##  [37] crosstalk_1.2.0             vroom_1.6.4                </span></span>
<span><span class="co">##  [39] car_3.1-2                   dplyr_1.1.4                </span></span>
<span><span class="co">##  [41] Matrix_1.6-1.1              RProtoBufLib_2.14.0        </span></span>
<span><span class="co">##  [43] ggbeeswarm_0.7.2            fansi_1.0.5                </span></span>
<span><span class="co">##  [45] abind_1.4-5                 terra_1.7-55               </span></span>
<span><span class="co">##  [47] lifecycle_1.0.4             multcomp_1.4-25            </span></span>
<span><span class="co">##  [49] yaml_2.3.7                  carData_3.0-5              </span></span>
<span><span class="co">##  [51] rhdf5_2.46.0                SparseArray_1.2.2          </span></span>
<span><span class="co">##  [53] Rtsne_0.16                  grid_4.3.2                 </span></span>
<span><span class="co">##  [55] promises_1.2.1              crayon_1.5.2               </span></span>
<span><span class="co">##  [57] shinydashboard_0.7.2        lattice_0.21-9             </span></span>
<span><span class="co">##  [59] beachmat_2.18.0             cowplot_1.1.1              </span></span>
<span><span class="co">##  [61] magick_2.8.1                pillar_1.9.0               </span></span>
<span><span class="co">##  [63] knitr_1.45                  ComplexHeatmap_2.18.0      </span></span>
<span><span class="co">##  [65] RTriangle_1.6-0.12          rjson_0.2.21               </span></span>
<span><span class="co">##  [67] codetools_0.2-19            glue_1.6.2                 </span></span>
<span><span class="co">##  [69] V8_4.4.0                    data.table_1.14.8          </span></span>
<span><span class="co">##  [71] vctrs_0.6.4                 png_0.1-8                  </span></span>
<span><span class="co">##  [73] gtable_0.3.4                cachem_1.0.8               </span></span>
<span><span class="co">##  [75] xfun_0.41                   S4Arrays_1.2.0             </span></span>
<span><span class="co">##  [77] mime_0.12                   tidygraph_1.2.3            </span></span>
<span><span class="co">##  [79] ConsensusClusterPlus_1.66.0 survival_3.5-7             </span></span>
<span><span class="co">##  [81] pheatmap_1.0.12             iterators_1.0.14           </span></span>
<span><span class="co">##  [83] cytolib_2.14.0              units_0.8-4                </span></span>
<span><span class="co">##  [85] ellipsis_0.3.2              TH.data_1.1-2              </span></span>
<span><span class="co">##  [87] bit64_4.0.5                 rprojroot_2.0.4            </span></span>
<span><span class="co">##  [89] bslib_0.5.1                 irlba_2.3.5.1              </span></span>
<span><span class="co">##  [91] svgPanZoom_0.3.4            vipor_0.4.5                </span></span>
<span><span class="co">##  [93] KernSmooth_2.23-22          colorspace_2.1-0           </span></span>
<span><span class="co">##  [95] DBI_1.1.3                   raster_3.6-26              </span></span>
<span><span class="co">##  [97] tidyselect_1.2.0            curl_5.1.0                 </span></span>
<span><span class="co">##  [99] bit_4.0.5                   compiler_4.3.2             </span></span>
<span><span class="co">## [101] BiocNeighbors_1.20.0        desc_1.4.2                 </span></span>
<span><span class="co">## [103] DelayedArray_0.28.0         bookdown_0.36              </span></span>
<span><span class="co">## [105] scales_1.2.1                classInt_0.4-10            </span></span>
<span><span class="co">## [107] distances_0.1.9             tiff_0.1-11                </span></span>
<span><span class="co">## [109] stringr_1.5.1               digest_0.6.33              </span></span>
<span><span class="co">## [111] fftwtools_0.9-11            rmarkdown_2.25             </span></span>
<span><span class="co">## [113] XVector_0.42.0              htmltools_0.5.7            </span></span>
<span><span class="co">## [115] pkgconfig_2.0.3             jpeg_0.1-10                </span></span>
<span><span class="co">## [117] sparseMatrixStats_1.14.0    highr_0.10                 </span></span>
<span><span class="co">## [119] fastmap_1.1.1               rlang_1.1.2                </span></span>
<span><span class="co">## [121] GlobalOptions_0.1.2         htmlwidgets_1.6.2          </span></span>
<span><span class="co">## [123] shiny_1.8.0                 DelayedMatrixStats_1.24.0  </span></span>
<span><span class="co">## [125] farver_2.1.1                jquerylib_0.1.4            </span></span>
<span><span class="co">## [127] zoo_1.8-12                  jsonlite_1.8.7             </span></span>
<span><span class="co">## [129] BiocParallel_1.36.0         BiocSingular_1.18.0        </span></span>
<span><span class="co">## [131] RCurl_1.98-1.13             magrittr_2.0.3             </span></span>
<span><span class="co">## [133] scuttle_1.12.0              GenomeInfoDbData_1.2.11    </span></span>
<span><span class="co">## [135] Rhdf5lib_1.24.0             munsell_0.5.0              </span></span>
<span><span class="co">## [137] Rcpp_1.0.11                 ggnewscale_0.4.9           </span></span>
<span><span class="co">## [139] viridis_0.6.4               stringi_1.8.1              </span></span>
<span><span class="co">## [141] zlibbioc_1.48.0             MASS_7.3-60                </span></span>
<span><span class="co">## [143] plyr_1.8.9                  parallel_4.3.2             </span></span>
<span><span class="co">## [145] ggrepel_0.9.4               graphlayouts_1.0.2         </span></span>
<span><span class="co">## [147] splines_4.3.2               hms_1.1.3                  </span></span>
<span><span class="co">## [149] circlize_0.4.15             locfit_1.5-9.8             </span></span>
<span><span class="co">## [151] igraph_1.5.1                ggpubr_0.6.0               </span></span>
<span><span class="co">## [153] ggsignif_0.6.4              ScaledMatrix_1.10.0        </span></span>
<span><span class="co">## [155] reshape2_1.4.4              XML_3.99-0.15              </span></span>
<span><span class="co">## [157] drc_3.0-1                   evaluate_0.23              </span></span>
<span><span class="co">## [159] BiocManager_1.30.22         tzdb_0.4.0                 </span></span>
<span><span class="co">## [161] foreach_1.5.2               tweenr_2.0.2               </span></span>
<span><span class="co">## [163] httpuv_1.6.12               tidyr_1.3.0                </span></span>
<span><span class="co">## [165] purrr_1.0.2                 polyclip_1.10-6            </span></span>
<span><span class="co">## [167] clue_0.3-65                 ggforce_0.4.1              </span></span>
<span><span class="co">## [169] rsvd_1.0.5                  broom_1.0.5                </span></span>
<span><span class="co">## [171] xtable_1.8-4                e1071_1.7-13               </span></span>
<span><span class="co">## [173] rstatix_0.7.2               later_1.3.1                </span></span>
<span><span class="co">## [175] viridisLite_0.4.2           class_7.3-22               </span></span>
<span><span class="co">## [177] ragg_1.2.6                  tibble_3.2.1               </span></span>
<span><span class="co">## [179] memoise_2.0.1               FlowSOM_2.10.0             </span></span>
<span><span class="co">## [181] beeswarm_0.4.0              cluster_2.1.4              </span></span>
<span><span class="co">## [183] concaveman_1.1.0</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Angelo2014" class="csl-entry">
Angelo, Michael, Sean C. Bendall, Rachel Finck, Matthew B. Hale, Chuck
Hitzman, Alexander D. Borowsky, Richard M. Levenson, et al. 2014.
<span>“Multiplexed Ion Beam Imaging of Human Breast Tumors.”</span>
<em>Nature Medicine</em> 20 (4): 436–42.
</div>
<div id="ref-Bhate2022" class="csl-entry">
Bhate, Salil S., Graham L. Barlow, Christian M. Schürch, and Garry P.
Nolan. 2022. <span>“Tissue Schematics Map the Specialization of Immune
Tissue Motifs and Their Appropriation by Tumors.”</span> <em>Cell
Systems</em> 13 (2): 109–30.
</div>
<div id="ref-Bodenmiller2016" class="csl-entry">
Bodenmiller, Bernd. 2016. <span>“Multiplexed Epitope-Based Tissue
Imaging for Discovery and Healthcare Applications.”</span> <em>Cell
Systems</em> 2: 225–38.
</div>
<div id="ref-Chen2015" class="csl-entry">
Chen, Kok Hao, Alistair N. Boettiger, Jeffrey R. Moffitt, Siyuan Wang,
and Xiaowei Zhuang. 2015. <span>“Spatially Resolved, Highly Multiplexed
RNA Profiling in Single Cells.”</span> <em>Science</em> 348: aaa6090.
</div>
<div id="ref-Chevrier2017" class="csl-entry">
Chevrier, Stéphane, Helena L. Crowell, Vito R. T. Zanotelli, Stefanie
Engler, Mark D. Robinson, and Bernd Bodenmiller. 2017.
<span>“Compensation of Signal Spillover in Suspension and Imaging Mass
Cytometry.”</span> <em>Cell Systems</em> 6: 612–20.
</div>
<div id="ref-Giesen2014" class="csl-entry">
Giesen, Charlotte, Hao A. O. Wang, Denis Schapiro, Nevena Zivanovic,
Andrea Jacobs, Bodo Hattendorf, Peter J. Schüffler, et al. 2014.
<span>“Highly Multiplexed Imaging of Tumor Tissues with Subcellular
Resolution by Mass Cytometry.”</span> <em>Nature
<span>M</span>ethods</em> 11 (4): 417–22.
</div>
<div id="ref-Greenwald2021" class="csl-entry">
Greenwald, Noah F, Geneva Miller, Erick Moen, Alex Kong, Adam Kagel,
Christine Camacho, Brianna J Mcintosh, et al. 2021. <span>“Whole-Cell
Segmentation of Tissue Images with Human-Level Performance Using
Large-Scale Data Annotation and Deep Learning.”</span> <em>bioRxiv</em>,
1–29.
</div>
<div id="ref-Gut2018" class="csl-entry">
Gut, Gabriele, Markus D Herrmann, and Lucas Pelkmans. 2018.
<span>“Multiplexed Protein Maps Link Subcellular Organization to
Cellular States.”</span> <em>Science</em> 361: 1–13.
</div>
<div id="ref-Jackson2020" class="csl-entry">
Jackson, Hartland W., Jana R. Fischer, Vito R. T. Zanotelli, H. Raza
Ali, Robert Mechera, Savas D. Soysal, Holger Moch, et al. 2020.
<span>“The Single-Cell Pathology Landscape of Breast Cancer.”</span>
<em>Nature</em> 578 (7796): 615–20.
</div>
<div id="ref-Lin2018" class="csl-entry">
Lin, Jia-Ren, Benjamin Izar, Shu Wang, Clarence Yapp, Shaolin Mei, Parin
M. Shah, Sandro Santagata, and Peter K. Sorger. 2018. <span>“Highly
Multiplexed Immunofluorescence Imaging of Human Tissues and Tumors Using
t-CyCIF and Conventional Optical Microscopes.”</span> <em>eLife</em> 7:
1–46.
</div>
<div id="ref-Lubeck2014" class="csl-entry">
Lubeck, Eric, Ahmet F Coskun, Timur Zhiyentayev, Mubhij Ahmad, and Long
Cai. 2014. <span>“Single-Cell in Situ RNA Profiling by Sequential
Hybridization.”</span> <em>Nature Methods</em> 11: 360–61.
</div>
<div id="ref-Shapiro2017" class="csl-entry">
Schapiro, Denis, Hartland W Jackson, Swetha Raghuraman, Jana R Fischer,
Vito RT Zanotelli, Daniel Schulz, Charlotte Giesen, Raúl Catena,
Zsuzsanna Varga, and Bernd Bodenmiller. 2017. <span>“histoCAT: Analysis
of Cell Phenotypes and Interactions in Multiplex Image Cytometry
Data.”</span> <em>Nature Methods</em> 14: 873--876.
</div>
<div id="ref-Schulz2018" class="csl-entry">
Schulz, Daniel, Vito RT Zanotelli, Rana R Fischer, Denis Schapiro,
Stefanie Engler, Xiao-Kang Lun, Hartland W Jackson, and Bernd
Bodenmiller. 2018. <span>“Simultaneous Multiplexed Imaging of mRNA and
Proteins with Subcellular Resolution in Breast Cancer Tissue Samples by
Mass Cytometry.”</span> <em>Cell Systems</em> 6: 25–36.e5.
</div>
<div id="ref-Schurch2020" class="csl-entry">
Schürch, Christian M., Salil S. Bhate, Graham L. Barlow, Darci J.
Phillips, Luca Noti, Inti Zlobec, Pauline Chu, et al. 2020.
<span>“Coordinated Cellular Neighborhoods Orchestrate Antitumoral
Immunity at the Colorectal Cancer Invasive Front.”</span> <em>Cell</em>
182 (5): 1341–59.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nils Eling.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
